<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>SALMON &mdash; Snakemake Wrappers heads/Unofficial documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="SNPEFF_ANNOTATE" href="snpeff_annotate.html" />
    <link rel="prev" title="MUTECT2_SOMATIC" href="mutect2_somatic.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Snakemake Wrappers
          </a>
              <div class="version">
                heads/Unofficial
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Wrappers, Meta-Wrappers and Pipelines</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">Wrappers</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../meta-wrappers.html">Meta-Wrappers</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="bioinfokit.html">DESEQ2_POST_PROCESS</a></li>
<li class="toctree-l2"><a class="reference internal" href="bismark.html">BISKMARK</a></li>
<li class="toctree-l2"><a class="reference internal" href="bismark_multisample.html">BISKMARK_MULTISAMPLE</a></li>
<li class="toctree-l2"><a class="reference internal" href="bowtie2.html">BOWTIE2</a></li>
<li class="toctree-l2"><a class="reference internal" href="bwa_fixmate.html">BWA_FIXMATE</a></li>
<li class="toctree-l2"><a class="reference internal" href="bwa_mapping.html">BWA_MAPPING</a></li>
<li class="toctree-l2"><a class="reference internal" href="clusterprofiler.html">CLUSTERP PROFILER</a></li>
<li class="toctree-l2"><a class="reference internal" href="compress_index_vcf.html">COMPRESS_INDEX_VCF</a></li>
<li class="toctree-l2"><a class="reference internal" href="coverage_variant_region.html">COVERAGE_VARIANT_REGION</a></li>
<li class="toctree-l2"><a class="reference internal" href="dada2_pe.html">DADA2-PE</a></li>
<li class="toctree-l2"><a class="reference internal" href="dada2_se.html">DADA2-SE</a></li>
<li class="toctree-l2"><a class="reference internal" href="deseq2_post_process.html">DESEQ2_POST_PROCESS</a></li>
<li class="toctree-l2"><a class="reference internal" href="download_references.html">DOWNLOAD REFERENCE</a></li>
<li class="toctree-l2"><a class="reference internal" href="eacon_cytoscan.html">EACON CYTOSCAN</a></li>
<li class="toctree-l2"><a class="reference internal" href="eacon_oncoscan.html">EACON ONCOSCAN</a></li>
<li class="toctree-l2"><a class="reference internal" href="eacon_post_process.html">EACON POST PROCESS</a></li>
<li class="toctree-l2"><a class="reference internal" href="fastq_screen_indexer.html">FASTQ_SCREEN_INDEXER</a></li>
<li class="toctree-l2"><a class="reference internal" href="gatk_bqsr.html">GATK_BQSR</a></li>
<li class="toctree-l2"><a class="reference internal" href="index_datasets.html">INDEX_DATASETS</a></li>
<li class="toctree-l2"><a class="reference internal" href="maftools.html">MAFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="meta_caller.html">META_CALLER</a></li>
<li class="toctree-l2"><a class="reference internal" href="meta_caller_germline.html">META_CALLER_GERMLINE</a></li>
<li class="toctree-l2"><a class="reference internal" href="msi_sensor_pro.html">MSISENSOR_PRO</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutect2_germline.html">MUTECT2_GERMLINE</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutect2_somatic.html">MUTECT2_SOMATIC</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">SALMON</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a></li>
<li class="toctree-l3"><a class="reference internal" href="#used-wrappers">Used wrappers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#authors">Authors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="snpeff_annotate.html">SNPEFF_ANNOTATE</a></li>
<li class="toctree-l2"><a class="reference internal" href="snpsift.html">SNPSIFT</a></li>
<li class="toctree-l2"><a class="reference internal" href="somatic_tmb.html">SOMATIC_TMB</a></li>
<li class="toctree-l2"><a class="reference internal" href="star_arriba.html">STAR-ARRIBA</a></li>
<li class="toctree-l2"><a class="reference internal" href="star_mapping.html">STAR_MAPPING</a></li>
<li class="toctree-l2"><a class="reference internal" href="tximport_deseq2.html">TXIMPORT_DESEQ2</a></li>
<li class="toctree-l2"><a class="reference internal" href="varscan2_calling.html">VARSCAN2_CALLING</a></li>
<li class="toctree-l2"><a class="reference internal" href="varscan2_germline.html">VARSCAN2_GERMLINE</a></li>
<li class="toctree-l2"><a class="reference internal" href="varscan2_somatic.html">VARSCAN2_SOMATIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="vcf_post_process.html">VCF_POST_PROCESS</a></li>
<li class="toctree-l2"><a class="reference internal" href="wisecondorx.html">WISECONDORX</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../pipelines.html">BiGR Pipelines</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Snakemake Wrappers</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../meta-wrappers.html">Meta-Wrappers</a> &raquo;</li>
      <li>SALMON</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/meta-wrappers/salmon.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="salmon">
<span id="id1"></span><h1>SALMON<a class="headerlink" href="#salmon" title="Permalink to this headline"></a></h1>
<p>Pseudo map and quantify your reads over transcritome with <a class="reference external" href="https://salmon.readthedocs.io/en/latest/">Salmon</a></p>
<section id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline"></a></h2>
<p>This meta-wrapper can be used by integrating the following into your workflow:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">snakemake.utils</span> <span class="kn">import</span> <span class="n">min_version</span>
<span class="n">min_version</span><span class="p">(</span><span class="s2">&quot;6.0&quot;</span><span class="p">)</span>

<span class="n">default_salmon_config</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;genome&quot;</span><span class="p">:</span> <span class="s2">&quot;path/to/sequence.fasta&quot;</span><span class="p">,</span>
    <span class="s2">&quot;transcriptome&quot;</span><span class="p">:</span> <span class="s2">&quot;path/to/transcriptome.fasta&quot;</span>
<span class="p">}</span>

<span class="k">def</span> <span class="nf">get_best_index</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Return cached indexes if available&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;salmon_index&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;salmon_index&quot;</span><span class="p">]</span>

    <span class="n">args</span> <span class="o">=</span> <span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;transcriptome&quot;</span><span class="p">],</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">],</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;salmon_index_extra&quot;</span><span class="p">])</span>
    <span class="n">hg38_flamingo</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh38.99/GRCh38.99.homo_sapiens.cdna.fasta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh38.99/GRCh38.99.homo_sapiens.dna.main_chr.fasta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--keepDuplicates --gencode&quot;</span>
    <span class="p">)</span>
    <span class="n">hg19_flamingo</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh37.75/GRCh37.75.homo_sapiens.cdna.fasta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh37.75/Homo_sapiens.GRCh37.75.dna.primary_assembly.main_chr.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--keepDuplicates --gencode&quot;</span>
    <span class="p">)</span>
    <span class="n">mm10_flamingo</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCm38.99/GRCm38.99.mus_musculus.cdna.nopatch.fasta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCm38.99/GRCm38.99.mus_musculus.dna.fasta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--keepDuplicates --gencode&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span> <span class="o">==</span> <span class="n">hg38_flamingo</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;salmon/index_hg38&quot;</span>
    <span class="k">elif</span> <span class="n">args</span> <span class="o">==</span> <span class="n">hg19_flamingo</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;salmon/index_hg19&quot;</span>
    <span class="k">elif</span> <span class="n">args</span> <span class="o">==</span> <span class="n">mm10_flamingo</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;salmon/index_mm10&quot;</span>

    <span class="k">return</span> <span class="s2">&quot;salmon/index&quot;</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This rule pseudo-map and quantifies your paired reads over the indexed</span>
<span class="sd">reference.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">rule</span> <span class="n">salmon_quant_paired</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">r1</span><span class="o">=</span><span class="s2">&quot;fastp/trimmed/pe/</span><span class="si">{sample}</span><span class="s2">.1.fastq&quot;</span><span class="p">,</span>
        <span class="n">r2</span><span class="o">=</span><span class="s2">&quot;fastp/trimmed/pe/</span><span class="si">{sample}</span><span class="s2">.2.fastq&quot;</span><span class="p">,</span>
        <span class="n">index</span><span class="o">=</span><span class="n">get_best_index</span><span class="p">(),</span>
        <span class="n">gtf</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;gtf&quot;</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">quant</span><span class="o">=</span><span class="s2">&quot;salmon/pseudo_mapping/</span><span class="si">{sample}</span><span class="s2">/quant.sf&quot;</span><span class="p">,</span>
        <span class="n">lib</span><span class="o">=</span><span class="s2">&quot;salmon/pseudo_mapping/</span><span class="si">{sample}</span><span class="s2">/lib_format_counts.json&quot;</span><span class="p">,</span>
        <span class="n">mapping</span><span class="o">=</span><span class="n">temp</span><span class="p">(</span><span class="s2">&quot;salmon/bams/</span><span class="si">{sample}</span><span class="s2">.bam&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Quantifying </span><span class="si">{wildcards.sample}</span><span class="s2"> with Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;threads&quot;</span><span class="p">,</span> <span class="mi">20</span><span class="p">),</span> <span class="mi">20</span><span class="p">)</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="nb">input</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">45</span> <span class="o">*</span> <span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">size_mb</span> <span class="o">/</span> <span class="mi">10240</span><span class="p">),</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="nb">input</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">1024</span> <span class="o">*</span> <span class="mi">10</span> <span class="o">*</span> <span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">size_mb</span> <span class="o">/</span> <span class="mi">10240</span><span class="p">),</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">libType</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;salmon_libtype&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">),</span>
        <span class="n">extra</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="s2">&quot;salmon_quant_extra&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--numBootstraps 100 --validateMappings --gcBias --seqBias --posBias&quot;</span>
        <span class="p">)</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/quant/</span><span class="si">{sample}</span><span class="s2">.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/quant&quot;</span>


<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Index your transcriptome or gentrome file with Salmon in order to map your</span>
<span class="sd">reads against this reference.</span>

<span class="sd">These rules are cached since it should be used only once per reference genome.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">rule</span> <span class="n">salmon_index</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">sequences</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.fasta&quot;</span><span class="p">,</span>
        <span class="n">decoys</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/decoy.txt&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">index</span><span class="o">=</span><span class="n">temp</span><span class="p">(</span><span class="n">directory</span><span class="p">(</span><span class="s2">&quot;salmon/index&quot;</span><span class="p">))</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Indexing transcriptome/gentrome sequences with Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;threads&quot;</span><span class="p">,</span> <span class="mi">20</span><span class="p">),</span> <span class="mi">20</span><span class="p">)</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">,</span> <span class="nb">input</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">attempt</span> <span class="o">*</span> <span class="p">(</span><span class="mi">120</span> <span class="k">if</span> <span class="s2">&quot;decoys&quot;</span> <span class="ow">in</span> <span class="nb">input</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">else</span> <span class="mi">45</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">,</span> <span class="nb">input</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">attempt</span> <span class="o">*</span> <span class="p">(</span><span class="mi">25600</span> <span class="k">if</span> <span class="s2">&quot;decoys&quot;</span> <span class="ow">in</span> <span class="nb">input</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">else</span> <span class="mi">10240</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">extra</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;salmon_index_extra&quot;</span><span class="p">,</span> <span class="s2">&quot;--keepDuplicates --gencode&quot;</span><span class="p">)</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/index.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/index&quot;</span>


<span class="n">rule</span> <span class="n">salmon_index_hg38</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">sequences</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.hg38.fasta&quot;</span><span class="p">,</span>
        <span class="n">decoys</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.hg38.decoy&quot;</span><span class="p">,</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">index</span><span class="o">=</span><span class="n">directory</span><span class="p">(</span><span class="s2">&quot;salmon/index_hg38&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Indexing transcriptome/gentrome sequences with Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">20</span>
    <span class="n">cache</span><span class="p">:</span> <span class="kc">True</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="mi">120</span><span class="p">,</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="mi">25600</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">extra</span><span class="o">=</span><span class="s2">&quot;--keepDuplicates --gencode&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/index.hg38.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/index&quot;</span>


<span class="n">rule</span> <span class="n">salmon_index_hg19</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">sequences</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.hg19.fasta&quot;</span><span class="p">,</span>
        <span class="n">decoys</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.hg19.decoy&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">index</span><span class="o">=</span><span class="n">directory</span><span class="p">(</span><span class="s2">&quot;salmon/index_hg19&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Indexing transcriptome/gentrome sequences with Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">20</span>
    <span class="n">cache</span><span class="p">:</span> <span class="kc">True</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="mi">120</span><span class="p">,</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="mi">25600</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">extra</span><span class="o">=</span><span class="s2">&quot;--keepDuplicates --gencode&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/index.hg19.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/index&quot;</span>


<span class="n">rule</span> <span class="n">salmon_index_mm10</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">sequences</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.mm10.fasta&quot;</span><span class="p">,</span>
        <span class="n">decoys</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.mm10.decoy&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">index</span><span class="o">=</span><span class="n">directory</span><span class="p">(</span><span class="s2">&quot;salmon/index_mm10&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Indexing transcriptome/gentrome sequences with Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">20</span>
    <span class="n">cache</span><span class="p">:</span> <span class="kc">True</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="mi">120</span><span class="p">,</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="mi">25600</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">extra</span><span class="o">=</span><span class="s2">&quot;--keepDuplicates --gencode&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/index.mm10.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/index&quot;</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This rule is optional in case you want to use decoy sequences within your</span>
<span class="sd">transcriptome. See salmon documentation for more information.</span>

<span class="sd">These rules are cached since it should be used only once per reference genome.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">rule</span> <span class="n">salmon_generate_decoy_sequence</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">transcriptome</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;transcriptome&quot;</span><span class="p">],</span>
        <span class="n">genome</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">decoys</span><span class="o">=</span><span class="n">temp</span><span class="p">(</span><span class="s2">&quot;salmon/decoy/decoys.txt&quot;</span><span class="p">),</span>
        <span class="n">gentrome</span><span class="o">=</span><span class="n">temp</span><span class="p">(</span><span class="s2">&quot;salmon/decoy/gentrome.fasta&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Building gentrome and decoy sequences for Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">attempt</span> <span class="o">*</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">),</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">512</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/decoys.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/generate_decoy&quot;</span>


<span class="n">rule</span> <span class="n">salmon_generate_decoy_sequence_hg38</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">transcriptome</span><span class="o">=</span><span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh38.99/GRCh38.99.homo_sapiens.cdna.fasta&quot;</span><span class="p">,</span>
        <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh38.99/GRCh38.99.homo_sapiens.dna.main_chr.fasta&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">multiext</span><span class="p">(</span><span class="s2">&quot;salmon/decoy/gentrome.hg38&quot;</span><span class="p">,</span> <span class="s2">&quot;.decoy&quot;</span><span class="p">,</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Building gentrome and decoy sequences for Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="n">cache</span><span class="p">:</span> <span class="kc">True</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.hg38&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/decoys.hg38.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/generate_decoy&quot;</span>


<span class="n">rule</span> <span class="n">salmon_generate_decoy_sequence_hg19</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">transcriptome</span><span class="o">=</span><span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh37.75/GRCh37.75.homo_sapiens.cdna.fasta&quot;</span><span class="p">,</span>
        <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCh37.75/Homo_sapiens.GRCh37.75.dna.primary_assembly.main_chr.fa&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">multiext</span><span class="p">(</span><span class="s2">&quot;salmon/decoy/gentrome.hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;.decoy&quot;</span><span class="p">,</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Building gentrome and decoy sequences for Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="n">cache</span><span class="p">:</span> <span class="kc">True</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/decoys.hg19.log&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.hg19&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/generate_decoy&quot;</span>


<span class="n">rule</span> <span class="n">salmon_generate_decoy_sequence_mm10</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">transcriptome</span><span class="o">=</span><span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCm38.99/GRCm38.99.mus_musculus.cdna.nopatch.fasta&quot;</span><span class="p">,</span>
        <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;/mnt/beegfs/database/bioinfo/Index_DB/Fasta/Ensembl/GRCm38.99/GRCm38.99.mus_musculus.dna.fasta&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">multiext</span><span class="p">(</span><span class="s2">&quot;salmon/decoy/gentrome.mm10&quot;</span><span class="p">,</span> <span class="s2">&quot;.decoys&quot;</span><span class="p">,</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span> <span class="s2">&quot;Building gentrome and decoy sequences for Salmon&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="n">cache</span><span class="p">:</span> <span class="kc">True</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">time_min</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
        <span class="n">tmpdir</span><span class="o">=</span><span class="s2">&quot;tmp&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/salmon/decoys.mm10.log&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;salmon/decoy/gentrome.mm10&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/salmon/generate_decoy&quot;</span>
</pre></div>
</div>
<p>Note that input, output and log file paths can be chosen freely, as long as the dependencies between the rules remain as listed here.
For additional parameters in each individual wrapper, please refer to their corresponding documentation (see links below).</p>
<p>When running with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake --use-conda
</pre></div>
</div>
<p>the software dependencies will be automatically deployed into an isolated environment before execution.</p>
</section>
<section id="used-wrappers">
<h2>Used wrappers<a class="headerlink" href="#used-wrappers" title="Permalink to this headline"></a></h2>
<p>The following individual wrappers are used in this meta-wrapper:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../wrappers/salmon/generate_decoy.html#bio-salmon-generate-decoy"><span class="std std-ref">GENERATE DECOY</span></a></p></li>
<li><p><a class="reference internal" href="../wrappers/salmon/index.html#bio-salmon-index"><span class="std std-ref">SALMON_INDEX</span></a></p></li>
<li><p><a class="reference internal" href="../wrappers/salmon/quant.html#bio-salmon-quant"><span class="std std-ref">SALMON_QUANT</span></a></p></li>
</ul>
<p>Please refer to each wrapper in above list for additional configuration parameters and information about the executed code.</p>
</section>
<section id="authors">
<h2>Authors<a class="headerlink" href="#authors" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>Thibault Dayris</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="mutect2_somatic.html" class="btn btn-neutral float-left" title="MUTECT2_SOMATIC" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="snpeff_annotate.html" class="btn btn-neutral float-right" title="SNPEFF_ANNOTATE" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2016, The Snakemake team.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>