

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>TCN_GIS_AUTOSCORER &mdash; Snakemake Wrappers heads/Unofficial documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="EACON INSTALL" href="install.html" />
    <link rel="prev" title="EACON_DOWNLOAD_DATABASES" href="databases.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> Snakemake Wrappers
          

          
          </a>

          
            
            
              <div class="version">
                heads/Unofficial
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Wrappers and Meta-Wrappers</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../wrappers.html">Wrappers</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../BiGR.html">BIGR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Rsamtools.html">RSAMTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../SomaticSignatures.html">SOMATICSIGNATURES</a></li>
<li class="toctree-l2"><a class="reference internal" href="../VariantAnnotation.html">VARIANTANNOTATION</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adapterremoval.html">ADAPTERREMOVAL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../arriba.html">ARRIBA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../art.html">ART</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bamtools.html">BAMTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bcftools.html">BCFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bcl2fastq.html">BCF2FASTQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bedtools.html">BEDTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../benchmark.html">BENCHMARK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bioinfokit.html">BIOINFOKIT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../biomaRt.html">BIOMART</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bismark.html">BISMARK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bowtie2.html">BOWTIE2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../busco.html">BUSCO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bwa.html">BWA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bwa-mem2.html">BWA-MEM2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cairosvg.html">CAIROSVG</a></li>
<li class="toctree-l2"><a class="reference internal" href="../check_md5.html">CHECK MD5</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseeker.html">CHIPSEEKER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../clustalo.html">CLUSTALO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../clusterProfiler.html">CLUSTERPROFILER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../concatenate_fastq.html">CONCATENATE FASTQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cp.html">CP (BASH)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cpat.html">CPAT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cutadapt.html">CUTADAPT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../dada2.html">DADA2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deeptools.html">DEEPTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deepvariant.html">DEEPVARIANT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../delly.html">DELLY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deseq2.html">DESEQ2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../diamond.html">DIAMOND</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../eacon.html">EACON</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="annotate.html">EACON_ANNOTATE</a></li>
<li class="toctree-l3"><a class="reference internal" href="ascn.html">EACON ASCN</a></li>
<li class="toctree-l3"><a class="reference internal" href="cytoscan_process.html">EACON_CYTOSCAN_PROCESS</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases.html">EACON_DOWNLOAD_DATABASES</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">TCN_GIS_AUTOSCORER</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#example">Example</a></li>
<li class="toctree-l4"><a class="reference internal" href="#software-dependencies">Software dependencies</a></li>
<li class="toctree-l4"><a class="reference internal" href="#input-output">Input/Output</a></li>
<li class="toctree-l4"><a class="reference internal" href="#authors">Authors</a></li>
<li class="toctree-l4"><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="install.html">EACON INSTALL</a></li>
<li class="toctree-l3"><a class="reference internal" href="oncoscan_process.html">EACON_ONCOSCAN_PROCESS</a></li>
<li class="toctree-l3"><a class="reference internal" href="segment.html">EACON_SEGMENT</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../enhancedVolcano.html">ENHANCEDVOLCANO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ensembl.html">ENSEMBL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../epic.html">EPIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fastp.html">FASTP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fastq_screen.html">FASTQ_SCREEN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fastqc.html">FASTQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fgbio.html">FGBIO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../filtlong.html">FILTLONG</a></li>
<li class="toctree-l2"><a class="reference internal" href="../freebayes.html">FREEBAYES</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gatk.html">GATK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gatk3.html">GATK3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gdc-api.html">GDC-API</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gdc-client.html">GDC-CLIENT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gencode.html">GENCODE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genomepy.html">GENOMEPY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gridss.html">GRIDSS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gtf.html">GTF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hap.py.html">HAP.PY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hisat2.html">HISAT2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hmmer.html">HMMER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../homer.html">HOMER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../iRODS.html">IRODS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../igv-reports.html">IGV-REPORTS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../infernal.html">INFERNAL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../isoformSwitchAnalyseR.html">ISOFORMSWITCHANALYSER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../jannovar.html">JANNOVAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../kallisto.html">KALLISTO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../last.html">LAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lofreq.html">LOFREQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../macs2.html">MACS2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../maftools.html">MAFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mapdamage2.html">MAPDAMAGE2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../md5sum.html">MD5SUM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../minimap2.html">MINIMAP2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mixcr.html">MIXCR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../msisensor.html">MSISENSOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multiqc.html">MULTIQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nanosim-h.html">NANOSIM-H</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ngs-disambiguate.html">NGS-DISAMBIGUATE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../open-cravat.html">OPEN-CRAVAT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../optitype.html">OPTITYPE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../paladin.html">PALADIN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pandas.html">PANDAS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pbmm2.html">PBMM2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pcaExplorer.html">PCAEXPLORER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pear.html">PEAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pfam.html">PFAM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pheatmap.html">PHEATMAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../picard.html">PICARD</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pindel.html">PINDEL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plass.html">PLASS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../preseq.html">PRESEQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../primerclip.html">PRIMERCLIP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../prosolo.html">PROSOLO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ptrimmer.html">PTRIMMER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pyfastaq.html">PYFASTAQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rasusa.html">RASUSA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../razers3.html">RAZERS3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rbt.html">RBT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rebaler.html">REBALER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference.html">REFERENCE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../refgenie.html">REFGENIE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rubic.html">RUBIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../salmon.html">SALMON</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sambamba.html">SAMBAMBA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samtools.html">SAMTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../seaborn.html">SEABORN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sed.html">SED</a></li>
<li class="toctree-l2"><a class="reference internal" href="../seqtk.html">SEQTK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../shovill.html">SHOVILL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sickle.html">SICKLE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../simplifyenrichment.html">SIMPLIFYENRICHMENT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sleuth.html">SLEUTH</a></li>
<li class="toctree-l2"><a class="reference internal" href="../snp-mutator.html">SNP-MUTATOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../snpeff.html">SNPEFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../snpsift.html">SNPSIFT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sourmash.html">SOURMASH</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sra-tools.html">SRA-TOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../star.html">STAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../strelka.html">STRELKA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../strling.html">STRLING</a></li>
<li class="toctree-l2"><a class="reference internal" href="../subread.html">SUBREAD</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tabix.html">TABIX</a></li>
<li class="toctree-l2"><a class="reference internal" href="../transdecoder.html">TRANSDECODER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trim_galore.html">TRIM_GALORE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trimmomatic.html">TRIMMOMATIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trinity.html">TRINITY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tximeta.html">TXIMETA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tximport.html">TXIMPORT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ucsc.html">UCSC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../umis.html">UMIS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../unicycler.html">UNICYCLER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vardict.html">VARDICT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../variant_distribution.html">VARIANT_DISTRIBUTION</a></li>
<li class="toctree-l2"><a class="reference internal" href="../varscan.html">VARSCAN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vcftools.html">VCFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vembrane.html">VEMBRANE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vep.html">VEP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vg.html">VG</a></li>
<li class="toctree-l2"><a class="reference internal" href="../wgsim.html">WGSIM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../meta-wrappers.html">Meta-Wrappers</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Snakemake Wrappers</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../wrappers.html">Wrappers</a> &raquo;</li>
        
          <li><a href="../eacon.html">EACON</a> &raquo;</li>
        
      <li>TCN_GIS_AUTOSCORER</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/wrappers/eacon/instability.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tcn-gis-autoscorer">
<span id="bio-eacon-instability"></span><h1>TCN_GIS_AUTOSCORER<a class="headerlink" href="#tcn-gis-autoscorer" title="Permalink to this headline">¶</a></h1>
<p>Score genomic instability based on ASCN models</p>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>This wrapper can be used in the following way:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">test_eacon_genomic_instability</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">gamma_eval</span> <span class="o">=</span> <span class="n">multiext</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">/ASCAT/ASCN/</span><span class="si">{sample}</span><span class="s2">.gemmeEval&quot;</span><span class="p">,</span>
            <span class="s2">&quot;.png&quot;</span><span class="p">,</span>
            <span class="s2">&quot;.txt&quot;</span>
        <span class="p">)</span>
        <span class="n">gamma_results</span> <span class="o">=</span> <span class="n">expand</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">/ASCAT/ASCN/gamma</span><span class="si">{gammas}</span><span class="s2">/</span><span class="si">{sample}</span><span class="s2">.</span><span class="si">{files}</span><span class="s2">,</span>
            <span class="n">gammas</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="o">/</span><span class="mi">100</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">95</span><span class="p">,</span> <span class="mi">5</span><span class="p">)],</span>
            <span class="n">files</span><span class="o">=</span><span class="p">[</span>
                <span class="s2">&quot;.rawprofile.png&quot;</span><span class="p">,</span>
                <span class="s2">&quot;.Rorschach.clown.png&quot;</span><span class="p">,</span>
                <span class="s2">&quot;.TCNvsL2R.png&quot;</span><span class="p">,</span>
                <span class="s2">&quot;.cn&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_model.txt&quot;</span><span class="p">,</span>
                <span class="s2">&quot;.ASCAT.png&quot;</span><span class="p">,</span>
                <span class="s2">&quot;.ASCAT.RDS&quot;</span>
            <span class="p">]</span>
        <span class="p">)</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">/</span><span class="si">{sample}</span><span class="s2">_GIS_from_best_gamma.txt&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/eacon_genomic_instability.log&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">indir</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">&quot;</span>  <span class="c1"># Required!</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;0.72.0-512-gb940f0d55/bio/eacon/instability&quot;</span>
</pre></div>
</div>
<p>Note that input, output and log file paths can be chosen freely.
When running with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake --use-conda
</pre></div>
</div>
<p>the software dependencies will be automatically deployed into an isolated environment before execution.</p>
</div>
<div class="section" id="software-dependencies">
<h2>Software dependencies<a class="headerlink" href="#software-dependencies" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">r-devtools==2.3.2</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">r-foreach==1.5.1</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">r-dplyr==1.0.5</span></code></p></li>
</ul>
</div>
<div class="section" id="input-output">
<h2>Input/Output<a class="headerlink" href="#input-output" title="Permalink to this headline">¶</a></h2>
<p><strong>Input:</strong></p>
<ul class="simple">
<li><p>An EaCoN gamma evaluation</p></li>
</ul>
<p><strong>Output:</strong></p>
<ul class="simple">
<li><p>A TSV formatted instability table</p></li>
</ul>
</div>
<div class="section" id="authors">
<h2>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Thibault Dayris</p></li>
<li><p>Bastien Job</p></li>
</ul>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## This script generates a table containing GIS (genomic instability scores) from an EaCoN TCN output using the ASCAT segmenter.</span>

<span class="c1">## Requirements :</span>
<span class="c1">## - CRAN : optparse, devtools, foreach, dplyr</span>
<span class="c1">## - GITHUB : https://github.com/aoumess/chromosomes</span>

<span class="c1"># Sink the stderr and stdout to the snakemake log file</span>
<span class="c1"># https://stackoverflow.com/a/48173272</span>
<span class="n">log.file</span> <span class="o">&lt;-</span> <span class="nf">file</span><span class="p">(</span>
  <span class="n">base</span><span class="o">::</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">snakemake</span><span class="o">@</span><span class="n">log</span><span class="p">[[</span><span class="m">1</span><span class="p">]]),</span>
  <span class="n">open</span><span class="o">=</span><span class="s">&quot;wt&quot;</span>
<span class="p">);</span>
<span class="n">base</span><span class="o">::</span><span class="nf">sink</span><span class="p">(</span><span class="n">log.file</span><span class="p">);</span>
<span class="n">base</span><span class="o">::</span><span class="nf">sink</span><span class="p">(</span><span class="n">log.file</span><span class="p">,</span><span class="n">type</span><span class="o">=</span><span class="s">&quot;message&quot;</span><span class="p">);</span>

<span class="n">indir</span> <span class="o">&lt;-</span> <span class="n">base</span><span class="o">::</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">[[</span><span class="s">&quot;indir&quot;</span><span class="p">]]);</span>

<span class="n">genome</span> <span class="o">&lt;-</span> <span class="s">&quot;hg19&quot;</span><span class="p">;</span>
<span class="nf">if </span><span class="p">(</span><span class="s">&quot;genome&quot;</span> <span class="o">%in%</span> <span class="n">base</span><span class="o">::</span><span class="nf">names</span><span class="p">(</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">))</span> <span class="p">{</span>
  <span class="n">genome</span> <span class="o">&lt;-</span> <span class="n">base</span><span class="o">::</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">[[</span><span class="s">&quot;genome&quot;</span><span class="p">]]);</span>
<span class="p">}</span>

<span class="n">segmenter</span> <span class="o">&lt;-</span> <span class="s">&quot;ASCAT&quot;</span><span class="p">;</span>
<span class="nf">if </span><span class="p">(</span><span class="s">&quot;segmenter&quot;</span> <span class="o">%in%</span> <span class="n">base</span><span class="o">::</span><span class="nf">names</span><span class="p">(</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">))</span> <span class="p">{</span>
  <span class="n">genome</span> <span class="o">&lt;-</span> <span class="n">base</span><span class="o">::</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">[[</span><span class="s">&quot;segmenter&quot;</span><span class="p">]]);</span>
<span class="p">}</span>

<span class="c1">## Checking args</span>
<span class="nf">if </span><span class="p">(</span><span class="nf">is.null</span><span class="p">(</span><span class="n">indir</span><span class="p">))</span> <span class="nf">stop</span><span class="p">(</span>
  <span class="s">&quot;An EaCoN results directory is required !&quot;</span>
<span class="p">);</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="n">indir</span><span class="p">))</span> <span class="nf">stop</span><span class="p">(</span>
  <span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;Input directory &#39;&quot;</span><span class="p">,</span> <span class="n">infile</span><span class="p">,</span> <span class="s">&quot;&#39; could not be found !&quot;</span><span class="p">)</span>
<span class="p">);</span>

<span class="c1">## Checking if chromosomes is installed</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="s">&#39;chromosomes&#39;</span> <span class="o">%in%</span> <span class="nf">installed.packages</span><span class="p">()))</span> <span class="p">{</span>
  <span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="s">&#39;devtools&#39;</span> <span class="o">%in%</span> <span class="nf">installed.packages</span><span class="p">()))</span> <span class="p">{</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;devtools&#39;</span><span class="p">)</span>
  <span class="p">}</span>
  <span class="n">devtools</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&#39;aoumess/chromosomes&#39;</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1">## Loading chromosomes data</span>
<span class="nf">data</span><span class="p">(</span><span class="n">list</span> <span class="o">=</span> <span class="n">genome</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&quot;chromosomes&quot;</span><span class="p">,</span> <span class="n">envir</span> <span class="o">=</span> <span class="nf">environment</span><span class="p">());</span>


<span class="c1">## Loading the gammaEval file</span>
<span class="n">samplename</span> <span class="o">&lt;-</span> <span class="nf">basename</span><span class="p">(</span><span class="n">indir</span><span class="p">);</span>
<span class="nf">if </span><span class="p">(</span><span class="s">&quot;sample_name&quot;</span> <span class="o">%in%</span> <span class="n">base</span><span class="o">::</span><span class="nf">names</span><span class="p">(</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">))</span> <span class="p">{</span>
  <span class="n">samplename</span> <span class="o">&lt;-</span> <span class="n">base</span><span class="o">::</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">snakemake</span><span class="o">@</span><span class="n">params</span><span class="p">[[</span><span class="s">&quot;samplename&quot;</span><span class="p">]]);</span>
<span class="p">}</span>
<span class="n">tcnsegdir</span> <span class="o">&lt;-</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">indir</span><span class="p">,</span> <span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="n">segmenter</span><span class="p">,</span> <span class="s">&#39;/ASCN/&#39;</span><span class="p">);</span>
<span class="n">gE</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span>
    <span class="nf">paste0</span><span class="p">(</span><span class="n">tcnsegdir</span><span class="p">,</span> <span class="n">samplename</span><span class="p">,</span> <span class="s">&#39;.gammaEval.txt&#39;</span><span class="p">),</span>
    <span class="n">header</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
    <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot;\t&quot;</span>
<span class="p">);</span>


<span class="c1">## Looking for the optimal model</span>
<span class="n">gE</span><span class="o">$</span><span class="n">round.psi</span> <span class="o">&lt;-</span> <span class="nf">round</span><span class="p">(</span><span class="n">gE</span><span class="o">$</span><span class="n">psi</span><span class="p">)</span>
<span class="n">gE</span><span class="o">$</span><span class="n">round.psi.diff</span> <span class="o">&lt;-</span> <span class="nf">abs</span><span class="p">(</span><span class="n">gE</span><span class="o">$</span><span class="n">psi</span> <span class="o">-</span> <span class="n">gE</span><span class="o">$</span><span class="n">round.psi</span><span class="p">)</span>
<span class="n">gE</span><span class="o">$</span><span class="n">gof.rank</span> <span class="o">&lt;-</span> <span class="nf">rank</span><span class="p">(</span><span class="n">gE</span><span class="o">$</span><span class="n">GoF</span><span class="p">)</span>
<span class="n">gE</span><span class="o">$</span><span class="n">rpd.rank</span> <span class="o">&lt;-</span> <span class="nf">rank</span><span class="p">(</span><span class="n">gE</span><span class="o">$</span><span class="n">round.psi.diff</span><span class="p">)</span>
<span class="n">gE</span><span class="o">$</span><span class="n">score</span> <span class="o">&lt;-</span> <span class="n">gE</span><span class="o">$</span><span class="n">gof.rank</span> <span class="o">-</span> <span class="n">gE</span><span class="o">$</span><span class="n">rpd.rank</span>

<span class="n">best.gamma</span> <span class="o">&lt;-</span> <span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;%.2f&quot;</span><span class="p">,</span> <span class="n">gE</span><span class="o">$</span><span class="n">gamma</span><span class="p">[</span><span class="nf">which.max</span><span class="p">(</span><span class="n">gE</span><span class="o">$</span><span class="n">score</span><span class="p">)])</span>

<span class="c1">## Loading TCN-CBS file</span>
<span class="n">as.res</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span>
  <span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span>
    <span class="n">tcnsegdir</span><span class="p">,</span> <span class="s">&#39;/gamma&#39;</span><span class="p">,</span> <span class="n">best.gamma</span><span class="p">,</span> <span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="n">samplename</span><span class="p">,</span> <span class="s">&#39;.ASCN.ASCAT.RDS&#39;</span>
  <span class="p">)</span>
<span class="p">);</span>

<span class="c1">## Computing needed values</span>
<span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">TCN</span> <span class="o">&lt;-</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">nMajor</span> <span class="o">+</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">nMajor</span>
<span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">Width</span> <span class="o">&lt;-</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">endpos</span> <span class="o">-</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">startpos</span> <span class="o">+</span> <span class="m">1</span>

<span class="c1">## Computing scores</span>
<span class="n">ploidy</span> <span class="o">&lt;-</span> <span class="n">as.res</span><span class="o">$</span><span class="n">ploidy</span><span class="o">$</span><span class="n">ascat</span>
<span class="n">gof</span> <span class="o">&lt;-</span> <span class="n">as.res</span><span class="o">$</span><span class="n">goodnessOfFit</span>
<span class="c1">### Sum of segments lengths multiplied by the absolute difference to diploidy,</span>
<span class="c1">### divided by the genome length.</span>
<span class="n">SKOR1</span> <span class="o">&lt;-</span> <span class="nf">sum</span><span class="p">(</span>
  <span class="nf">abs</span><span class="p">(</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">TCN</span> <span class="o">-</span> <span class="m">2</span>
<span class="p">)</span> <span class="o">*</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">Width</span><span class="p">)</span> <span class="o">/</span> <span class="n">cs</span><span class="o">$</span><span class="n">genome.length</span>
<span class="c1">### Sum of segments lengths multiplied by the absolute difference</span>
<span class="c1">### to exact ASCAT2-predicted ploidy, divided by the genome length.</span>
<span class="n">SKOR2</span> <span class="o">&lt;-</span> <span class="nf">sum</span><span class="p">(</span>
  <span class="nf">abs</span><span class="p">(</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">TCN</span> <span class="o">-</span> <span class="n">ploidy</span>
<span class="p">)</span> <span class="o">*</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">Width</span><span class="p">)</span> <span class="o">/</span> <span class="n">cs</span><span class="o">$</span><span class="n">genome.length</span>
<span class="c1">### Sum of segments lengths multiplied by the absolute difference to rounded ASCAT2-predicted ploidy, divided by the genome length.</span>
<span class="n">SKOR3</span> <span class="o">&lt;-</span> <span class="nf">sum</span><span class="p">(</span>
  <span class="nf">abs</span><span class="p">(</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">TCN</span> <span class="o">-</span> <span class="nf">round</span><span class="p">(</span><span class="n">ploidy</span><span class="p">)</span>
<span class="p">)</span> <span class="o">*</span> <span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">Width</span><span class="p">)</span> <span class="o">/</span> <span class="n">cs</span><span class="o">$</span><span class="n">genome.length</span>
<span class="c1">### Nb of segments</span>
<span class="n">NBSEG</span> <span class="o">&lt;-</span> <span class="nf">nrow</span><span class="p">(</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="p">)</span>
<span class="c1">### Compute a score on a per-chromosome basis, taking the longest ACN (in bp)</span>
<span class="c1">### as normal basis</span>
<span class="nf">library</span><span class="p">(</span><span class="n">foreach</span><span class="p">)</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">))</span>
<span class="n">LOKAL1</span> <span class="o">&lt;-</span> <span class="nf">sum</span><span class="p">(</span><span class="nf">foreach </span><span class="p">(</span><span class="n">k</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">chr</span><span class="p">),</span> <span class="n">.combine</span> <span class="o">=</span> <span class="s">&quot;c&quot;</span><span class="p">)</span> <span class="o">%do%</span> <span class="p">{</span>
  <span class="n">miniseg</span> <span class="o">&lt;-</span> <span class="n">tibble</span><span class="o">::</span><span class="nf">as_tibble</span><span class="p">(</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="p">[</span><span class="n">as.res</span><span class="o">$</span><span class="n">segments</span><span class="o">$</span><span class="n">chr</span> <span class="o">==</span> <span class="n">k</span><span class="p">,])</span>
  <span class="n">basistbl</span> <span class="o">&lt;-</span> <span class="n">miniseg</span> <span class="o">%&gt;%</span> <span class="nf">group_by</span><span class="p">(</span><span class="n">TCN</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">summarise</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">Width</span><span class="p">))</span>
  <span class="n">basis</span> <span class="o">&lt;-</span> <span class="n">basistbl</span><span class="o">$</span><span class="n">TCN</span><span class="p">[</span><span class="nf">which.max</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">basistbl</span><span class="p">)[,</span><span class="m">2</span><span class="p">])]</span>
  <span class="n">KSCOR</span> <span class="o">&lt;-</span> <span class="nf">sum</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">miniseg</span><span class="o">$</span><span class="n">TCN</span>  <span class="o">-</span> <span class="n">basis</span><span class="p">)</span> <span class="o">*</span> <span class="n">miniseg</span><span class="o">$</span><span class="n">Width</span><span class="p">)</span> <span class="o">/</span> <span class="n">cs</span><span class="o">$</span><span class="n">genome.length</span>
  <span class="nf">return</span><span class="p">(</span><span class="n">KSCOR</span><span class="p">)</span>
<span class="p">})</span>

<span class="c1">## Generating output</span>
<span class="n">outdf</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span>
  <span class="n">best.gamma</span> <span class="o">=</span> <span class="n">best.gamma</span><span class="p">,</span>
  <span class="n">psi</span> <span class="o">=</span> <span class="n">gE</span><span class="p">[</span><span class="n">best.gamma</span><span class="p">,</span> <span class="n">psi</span><span class="p">],</span>
  <span class="n">gof</span> <span class="o">=</span> <span class="n">gE</span><span class="p">[</span><span class="n">best.gamma</span><span class="p">,</span> <span class="n">gof</span><span class="p">],</span>
  <span class="n">score</span> <span class="o">=</span> <span class="n">gE</span><span class="p">[</span><span class="n">best.gamma</span><span class="p">,</span> <span class="n">score</span><span class="p">],</span>
  <span class="n">SCORE1</span> <span class="o">=</span> <span class="n">SKOR1</span><span class="p">,</span>
  <span class="n">SCORE2</span> <span class="o">=</span> <span class="n">SKOR2</span><span class="p">,</span>
  <span class="n">SCORE3</span> <span class="o">=</span> <span class="n">SKOR3</span><span class="p">,</span>
  <span class="n">NBSEG</span> <span class="o">=</span> <span class="n">NBSEG</span><span class="p">,</span>
  <span class="n">LOCAL1</span> <span class="o">=</span> <span class="n">LOKAL1</span><span class="p">,</span>
  <span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="kc">FALSE</span>
<span class="p">);</span>
<span class="nf">write.table</span><span class="p">(</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">outdf</span><span class="p">,</span>
    <span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span>
      <span class="n">indir</span><span class="p">,</span> <span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="n">samplename</span><span class="p">,</span> <span class="s">&#39;_GIS_from_best_gamma.txt&#39;</span>
    <span class="p">),</span>
    <span class="n">quote</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
    <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">,</span>
    <span class="n">row.names</span> <span class="o">=</span> <span class="kc">FALSE</span>
<span class="p">);</span>

<span class="c1"># Proper syntax to close the connection for the log file</span>
<span class="c1"># but could be optional for Snakemake wrapper</span>
<span class="n">base</span><span class="o">::</span><span class="nf">sink</span><span class="p">(</span><span class="n">type</span><span class="o">=</span><span class="s">&quot;message&quot;</span><span class="p">);</span>
<span class="n">base</span><span class="o">::</span><span class="nf">sink</span><span class="p">();</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="install.html" class="btn btn-neutral float-right" title="EACON INSTALL" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="databases.html" class="btn btn-neutral float-left" title="EACON_DOWNLOAD_DATABASES" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2016, The Snakemake team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>