

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>RENAME_SNPSIFT_MAF_COLS &mdash; Snakemake Wrappers heads/Unofficial documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="SPLIT_DESIGN" href="split_design.html" />
    <link rel="prev" title="MULTIQC_RNASEQ_REPORT" href="multiqc_rnaseq_report.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> Snakemake Wrappers
          

          
          </a>

          
            
            
              <div class="version">
                heads/Unofficial
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Wrappers, Meta-Wrappers and Pipelines</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../wrappers.html">Wrappers</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../BiGR.html">BIGR</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="copy.html">IGR_COPY</a></li>
<li class="toctree-l3"><a class="reference internal" href="drop_duplicates_md5.html">DROP_MD5_DUPLICATES</a></li>
<li class="toctree-l3"><a class="reference internal" href="fix_vcf.html">FIX VCF</a></li>
<li class="toctree-l3"><a class="reference internal" href="multiqc_rnaseq_report.html">MULTIQC_RNASEQ_REPORT</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">RENAME_SNPSIFT_MAF_COLS</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#example">Example</a></li>
<li class="toctree-l4"><a class="reference internal" href="#software-dependencies">Software dependencies</a></li>
<li class="toctree-l4"><a class="reference internal" href="#input-output">Input/Output</a></li>
<li class="toctree-l4"><a class="reference internal" href="#params">Params</a></li>
<li class="toctree-l4"><a class="reference internal" href="#notes">Notes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#authors">Authors</a></li>
<li class="toctree-l4"><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="split_design.html">SPLIT_DESIGN</a></li>
<li class="toctree-l3"><a class="reference internal" href="split_vcf_features.html">SPLIT VCF FEATURES</a></li>
<li class="toctree-l3"><a class="reference internal" href="split_vcf_multiallelic.html">SPLIT VCF MULTIALLELIC</a></li>
<li class="toctree-l3"><a class="reference internal" href="vcf_format_to_info.html">VCF_FORMAT_TO_INFO</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../Rsamtools.html">RSAMTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../SomaticSignatures.html">SOMATICSIGNATURES</a></li>
<li class="toctree-l2"><a class="reference internal" href="../VariantAnnotation.html">VARIANTANNOTATION</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adapterremoval.html">ADAPTERREMOVAL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../arriba.html">ARRIBA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../art.html">ART</a></li>
<li class="toctree-l2"><a class="reference internal" href="../assembly-stats.html">ASSEMBLY-STATS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bamtools.html">BAMTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bbtools.html">BBTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bcftools.html">BCFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bcl2fastq.html">BCF2FASTQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bedtools.html">BEDTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../benchmark.html">BENCHMARK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../biobambam2.html">BIOBAMBAM2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bioinfokit.html">BIOINFOKIT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../biomaRt.html">BIOMART</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bismark.html">BISMARK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../blast.html">BLAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bowtie2.html">BOWTIE2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../busco.html">BUSCO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bwa.html">BWA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../bwa-mem2.html">BWA-MEM2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cairosvg.html">CAIROSVG</a></li>
<li class="toctree-l2"><a class="reference internal" href="../check_md5.html">CHECK MD5</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseeker.html">CHIPSEEKER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../clustalo.html">CLUSTALO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../clusterProfiler.html">CLUSTERPROFILER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../compress.html">COMPRESS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../concatenate_fastq.html">CONCATENATE FASTQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../consensusclusterplus.html">CONSENSUSCLUSTERPLUS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cp.html">CP (BASH)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cpat.html">CPAT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cutadapt.html">CUTADAPT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../dada2.html">DADA2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deeptools.html">DEEPTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deepvariant.html">DEEPVARIANT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../delly.html">DELLY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deseq2.html">DESEQ2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../diamond.html">DIAMOND</a></li>
<li class="toctree-l2"><a class="reference internal" href="../eacon.html">EACON</a></li>
<li class="toctree-l2"><a class="reference internal" href="../enhancedVolcano.html">ENHANCEDVOLCANO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ensembl.html">ENSEMBL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../epic.html">EPIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../facets.html">FACETS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fastp.html">FASTP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fastq_screen.html">FASTQ_SCREEN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fastqc.html">FASTQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fgbio.html">FGBIO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fgsea.html">FGSEA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../filtlong.html">FILTLONG</a></li>
<li class="toctree-l2"><a class="reference internal" href="../freebayes.html">FREEBAYES</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gatk.html">GATK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gatk3.html">GATK3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gdc-api.html">GDC-API</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gdc-client.html">GDC-CLIENT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gencode.html">GENCODE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genomepy.html">GENOMEPY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gridss.html">GRIDSS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gsva.html">GSVA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gtf.html">GTF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hap.py.html">HAP.PY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hisat2.html">HISAT2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hmmer.html">HMMER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../homer.html">HOMER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../iRODS.html">IRODS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../igv-reports.html">IGV-REPORTS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../immunedeconv.html">IMMUNEDECONV</a></li>
<li class="toctree-l2"><a class="reference internal" href="../infernal.html">INFERNAL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../isoformSwitchAnalyseR.html">ISOFORMSWITCHANALYSER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../jannovar.html">JANNOVAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../kallisto.html">KALLISTO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../last.html">LAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="../liftoff.html">LIFTOFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lofreq.html">LOFREQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../macs2.html">MACS2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../maftools.html">MAFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mapdamage2.html">MAPDAMAGE2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../md5sum.html">MD5SUM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../minimap2.html">MINIMAP2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mixcr.html">MIXCR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mlst.html">MLST</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mosdepth.html">MOSDEPTH</a></li>
<li class="toctree-l2"><a class="reference internal" href="../msisensor.html">MSISENSOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multiqc.html">MULTIQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nanosim-h.html">NANOSIM-H</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">NEXTFLOW</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ngs-disambiguate.html">NGS-DISAMBIGUATE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../open-cravat.html">OPEN-CRAVAT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../optitype.html">OPTITYPE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../paladin.html">PALADIN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pandas.html">PANDAS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pandora.html">PANDORA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pbmm2.html">PBMM2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pcaExplorer.html">PCAEXPLORER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pear.html">PEAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pfam.html">PFAM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pheatmap.html">PHEATMAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../picard.html">PICARD</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pindel.html">PINDEL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plass.html">PLASS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../preseq.html">PRESEQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../primerclip.html">PRIMERCLIP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../prosolo.html">PROSOLO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ptrimmer.html">PTRIMMER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pyfastaq.html">PYFASTAQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../qualimap.html">QUALIMAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rasusa.html">RASUSA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../razers3.html">RAZERS3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rbt.html">RBT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rebaler.html">REBALER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference.html">REFERENCE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../refgenie.html">REFGENIE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rsem.html">RSEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rubic.html">RUBIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../salmon.html">SALMON</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sambamba.html">SAMBAMBA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../samtools.html">SAMTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../seaborn.html">SEABORN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sed.html">SED</a></li>
<li class="toctree-l2"><a class="reference internal" href="../seqtk.html">SEQTK</a></li>
<li class="toctree-l2"><a class="reference internal" href="../shovill.html">SHOVILL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sickle.html">SICKLE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../simplifyenrichment.html">SIMPLIFYENRICHMENT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sleuth.html">SLEUTH</a></li>
<li class="toctree-l2"><a class="reference internal" href="../snp-mutator.html">SNP-MUTATOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../snpeff.html">SNPEFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../snpsift.html">SNPSIFT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sourmash.html">SOURMASH</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sra-tools.html">SRA-TOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../star.html">STAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../strelka.html">STRELKA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../strling.html">STRLING</a></li>
<li class="toctree-l2"><a class="reference internal" href="../subread.html">SUBREAD</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tabix.html">TABIX</a></li>
<li class="toctree-l2"><a class="reference internal" href="../transdecoder.html">TRANSDECODER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trim_galore.html">TRIM_GALORE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trimmomatic.html">TRIMMOMATIC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trinity.html">TRINITY</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tximeta.html">TXIMETA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tximport.html">TXIMPORT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ucsc.html">UCSC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../umis.html">UMIS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../unicycler.html">UNICYCLER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vardict.html">VARDICT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../variant_distribution.html">VARIANT_DISTRIBUTION</a></li>
<li class="toctree-l2"><a class="reference internal" href="../variantoccurence.html">VARIANTOCCURENCE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../varscan.html">VARSCAN</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vcftools.html">VCFTOOLS</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vembrane.html">VEMBRANE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vep.html">VEP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../vg.html">VG</a></li>
<li class="toctree-l2"><a class="reference internal" href="../wgsim.html">WGSIM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../meta-wrappers.html">Meta-Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pipelines.html">BiGR Pipelines</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Snakemake Wrappers</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../wrappers.html">Wrappers</a> &raquo;</li>
        
          <li><a href="../BiGR.html">BIGR</a> &raquo;</li>
        
      <li>RENAME_SNPSIFT_MAF_COLS</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/wrappers/BiGR/rename_snpsift_maf_cols.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rename-snpsift-maf-cols">
<span id="bio-bigr-rename-snpsift-maf-cols"></span><h1>RENAME_SNPSIFT_MAF_COLS<a class="headerlink" href="#rename-snpsift-maf-cols" title="Permalink to this headline">¶</a></h1>
<p>Columns from snpsift are usually named in a non-human readable manner.</p>
<p><strong>URL</strong>:</p>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>This wrapper can be used in the following way:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">test_rename_snpsift_maf_cols</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">tsv</span> <span class="o">=</span> <span class="s2">&quot;mouse.tsv&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">tsv</span> <span class="o">=</span> <span class="s2">&quot;mouse.renamed.tsv&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/renamed.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;0.77.0-1111-g021ee08d2/bio/BiGR/rename_snpsift_maf_cols&quot;</span>
</pre></div>
</div>
<p>Note that input, output and log file paths can be chosen freely.</p>
<p>When running with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake --use-conda
</pre></div>
</div>
<p>the software dependencies will be automatically deployed into an isolated environment before execution.</p>
</div>
<div class="section" id="software-dependencies">
<h2>Software dependencies<a class="headerlink" href="#software-dependencies" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">python==3.9.6</span></code></p></li>
</ul>
</div>
<div class="section" id="input-output">
<h2>Input/Output<a class="headerlink" href="#input-output" title="Permalink to this headline">¶</a></h2>
<p><strong>Input:</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">tsv</span></code>: TSV formatted calls</p></li>
</ul>
<p><strong>Output:</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">tsv</span></code>: TSV formatted calls, with renamed columns</p></li>
</ul>
</div>
<div class="section" id="params">
<h2>Params<a class="headerlink" href="#params" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">add_cols</span></code>: Boolean. Add (default) optional columns, or don’t.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ncbi_build</span></code>: The NCBI build version of the genome. Required if and only if add_cols is True.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">center</span></code>: Sequencing center. Required if and only if add_cols is True.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">caller</span></code>: Used variant caller. One of <cite>Mutect2</cite>. Required if and only if add_cols is True.</p></li>
</ul>
</div>
<div class="section" id="notes">
<h2>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h2>
<p>Further columns descriptions may be added.</p>
</div>
<div class="section" id="authors">
<h2>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Thibault Dayris</p></li>
</ul>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/python3.8</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Rename splitted vcf columns</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
    <span class="n">filename</span><span class="o">=</span><span class="n">snakemake</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">filemode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span>
    <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span>
<span class="p">)</span>

<span class="k">def</span> <span class="nf">get_variant_type</span><span class="p">(</span><span class="n">ref</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">alt</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return variant type given the reference and the alternative alleles</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span> <span class="c1"># Nothing in REF, added in ALT</span>
        <span class="k">return</span> <span class="s2">&quot;INS&quot;</span>
    <span class="k">if</span> <span class="n">alt</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;DEL&quot;</span> <span class="c1"># Something in REF, nothing in ALT</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;SNP&quot;</span> <span class="c1"># REF and ALT only have one nucleotide</span>
        <span class="k">return</span> <span class="s2">&quot;ONP&quot;</span> <span class="c1"># REF had less nucleotides than ALT</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;DNP&quot;</span> <span class="c1"># REF and ALT have two nucleotides</span>
        <span class="k">return</span> <span class="s2">&quot;ONP&quot;</span> <span class="c1"># REF has two nucleotides, ALT have unknown</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;TNP&quot;</span> <span class="c1"># REF and ALT have three nucleotides</span>
        <span class="k">return</span> <span class="s2">&quot;ONP&quot;</span> <span class="c1"># REF has three nucleotides, ALT have unknown</span>
    <span class="k">return</span> <span class="s2">&quot;ONP&quot;</span> <span class="c1"># More than 3 nucleotides in ref, unknown in ALT</span>


<span class="n">headers</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;CHROM&quot;</span><span class="p">:</span> <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span>
    <span class="s2">&quot;POS&quot;</span><span class="p">:</span> <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_ID&quot;</span><span class="p">,</span>
    <span class="s2">&quot;REF3&quot;</span><span class="p">:</span> <span class="s2">&quot;Reference_Allele&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ALT&quot;</span><span class="p">:</span> <span class="s2">&quot;Tumor_Seq_Allele1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FILTER&quot;</span><span class="p">:</span> <span class="s2">&quot;Filter&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:AD&quot;</span><span class="p">:</span> <span class="s2">&quot;Allelic_depths&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:AF&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_fractions&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:DP&quot;</span><span class="p">:</span> <span class="s2">&quot;Pileup_Read_Depth&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:F1R2&quot;</span><span class="p">:</span> <span class="s2">&quot;F1R2_orientation&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:F2R1&quot;</span><span class="p">:</span> <span class="s2">&quot;F2R1_orientation&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:GQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Genotype_quality&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:GT&quot;</span><span class="p">:</span> <span class="s2">&quot;Genotype&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:PGT&quot;</span><span class="p">:</span> <span class="s2">&quot;Physical_phasing_haplotype&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:PID&quot;</span><span class="p">:</span> <span class="s2">&quot;Physical_phasing_ID&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:PL&quot;</span><span class="p">:</span> <span class="s2">&quot;Normalized_Phred_scaled_likelihoods&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:PS&quot;</span><span class="p">:</span> <span class="s2">&quot;Phasing_set&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FORMAT:SB&quot;</span><span class="p">:</span> <span class="s2">&quot;Strand_bias_Fisher_component&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_SB_TABLE&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_specific_Strand_bias&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_UNIQ_ALT_READ_COUNT&quot;</span><span class="p">:</span> <span class="s2">&quot;Unique_alt_variant_count&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CONTQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred_scaled_qualities_non_contamination&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DP&quot;</span><span class="p">:</span> <span class="s2">&quot;Mutect2_Read_depth&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ECNT&quot;</span><span class="p">:</span> <span class="s2">&quot;Events_in_haplotype&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GERMQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred_scaled_quality_non_germline&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MBQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Median_base_quality&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MFRL&quot;</span><span class="p">:</span> <span class="s2">&quot;Median_fragment_length&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MMQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Median_mapping_quality&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MPOS&quot;</span><span class="p">:</span> <span class="s2">&quot;Median_distance_to_end_read&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NALOD&quot;</span><span class="p">:</span> <span class="s2">&quot;Negative_log_10_odds_of_artifact&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NCount&quot;</span><span class="p">:</span> <span class="s2">&quot;Pileup_N_base_depth&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NLOD&quot;</span><span class="p">:</span> <span class="s2">&quot;Normal_log_10_likelihood_ploïdy&quot;</span><span class="p">,</span>
    <span class="s2">&quot;OCM&quot;</span><span class="p">:</span> <span class="s2">&quot;Non_matching_original_alignment&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PON&quot;</span><span class="p">:</span> <span class="s2">&quot;Panel_of_Normal&quot;</span><span class="p">,</span>
    <span class="s2">&quot;POPAF&quot;</span><span class="p">:</span> <span class="s2">&quot;negative_log_10_population_allele_frequencies&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ROQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred_scaled_qualities_not_orientation_bias&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RPA&quot;</span><span class="p">:</span> <span class="s2">&quot;Number_tandem_repetition&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RU&quot;</span><span class="p">:</span> <span class="s2">&quot;Repeat_unit&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SEQQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred_scaled_quality_not_sequencing_error&quot;</span><span class="p">,</span>
    <span class="s2">&quot;STR&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_short_tandem_repeat&quot;</span><span class="p">,</span>
    <span class="s2">&quot;STRANDQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred_scaled_quality_not_strand_bias&quot;</span><span class="p">,</span>
    <span class="s2">&quot;STRQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred_scaled_quality_not_tandem_polymerase_error&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TLOD&quot;</span><span class="p">:</span> <span class="s2">&quot;Log_10_likelihood_variant_exists&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].ALLELE&quot;</span><span class="p">:</span> <span class="s2">&quot;SnpEff_Genotype&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].EFFECT&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_Classification&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].IMPACT&quot;</span><span class="p">:</span> <span class="s2">&quot;IMPACT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].GENE&quot;</span><span class="p">:</span> <span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].GENEID&quot;</span><span class="p">:</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].FEATURE&quot;</span><span class="p">:</span> <span class="s2">&quot;Feature&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].FEATUREID&quot;</span><span class="p">:</span> <span class="s2">&quot;Transcript_ID&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].BIOTYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;BIOTYPE&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].RANK&quot;</span><span class="p">:</span> <span class="s2">&quot;Exon_Intron_rank&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].HGVS_C&quot;</span><span class="p">:</span> <span class="s2">&quot;HGVSc&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].HGVS_P&quot;</span><span class="p">:</span> <span class="s2">&quot;HGVSp&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].CDNA_POS&quot;</span><span class="p">:</span> <span class="s2">&quot;cDNA_position&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].CDNA_LENANN[*].CDS_POS&quot;</span><span class="p">:</span> <span class="s2">&quot;CDS_position&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].CDS_LEN&quot;</span><span class="p">:</span> <span class="s2">&quot;CDS_length&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].AA_POS&quot;</span><span class="p">:</span> <span class="s2">&quot;Protein_position&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].AA_LEN&quot;</span><span class="p">:</span> <span class="s2">&quot;Protein_length&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].DISTANCE&quot;</span><span class="p">:</span> <span class="s2">&quot;DISTANCE&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ANN[*].ERRORS&quot;</span><span class="p">:</span> <span class="s2">&quot;Errors&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LOF&quot;</span><span class="p">:</span> <span class="s2">&quot;Loss_of_Function_SnpEff&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NMD&quot;</span><span class="p">:</span> <span class="s2">&quot;Nonsense_mediated_decay&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VARTYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_types&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SNP&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_SNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MNP&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_MNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;INS&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_INS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DEL&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_DEL&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MIXED&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_MIXED_Polymorphisms&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HOM&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_Homozygous&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HET&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_Heterozygous&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MSigDb&quot;</span><span class="p">:</span> <span class="s2">&quot;MSigDb_Pathway&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Count&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DS&quot;</span><span class="p">:</span> <span class="s2">&quot;Database_Source&quot;</span><span class="p">,</span>
    <span class="s2">&quot;END&quot;</span><span class="p">:</span> <span class="s2">&quot;End_Position&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CDA&quot;</span><span class="p">:</span> <span class="s2">&quot;Clinical_Diagnostic_Assay&quot;</span><span class="p">,</span>
    <span class="s2">&quot;OTH&quot;</span><span class="p">:</span> <span class="s2">&quot;Orthologous_Variants_in_NCBI&quot;</span><span class="p">,</span>
    <span class="s2">&quot;S3D&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_known_3D_structure&quot;</span><span class="p">,</span>
    <span class="s2">&quot;WTD&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_Withdrawn_by_submitter&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbSNPBuildID&quot;</span><span class="p">:</span> <span class="s2">&quot;dbSNP_id&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SLO&quot;</span><span class="p">:</span> <span class="s2">&quot;SubmitterLinkOut&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NSF&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_Non_Synonymous_Frameshift&quot;</span><span class="p">,</span>
    <span class="s2">&quot;R3&quot;</span><span class="p">:</span> <span class="s2">&quot;3p_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;R5&quot;</span><span class="p">:</span> <span class="s2">&quot;5p_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NSN&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_Nonsense_Changes_Stop&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NSM&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_Nonsense_Changes_Peptide&quot;</span><span class="p">,</span>
    <span class="s2">&quot;G5A&quot;</span><span class="p">:</span> <span class="s2">&quot;Less_5pct_population&quot;</span><span class="p">,</span>
    <span class="s2">&quot;COMMON&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_common_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RS&quot;</span><span class="p">:</span> <span class="s2">&quot;RS_id_dbSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RV&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_is_reversed&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TPA&quot;</span><span class="p">:</span> <span class="s2">&quot;Third_party_annotation&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CFL&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_assembly_conflict&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GNO&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_Available_genotype&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VLD&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_Validated&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ASP&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_Assembly_Specific&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ASS&quot;</span><span class="p">:</span> <span class="s2">&quot;Acceptor_splice_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;REF&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_identical_to_reference&quot;</span><span class="p">,</span>
    <span class="s2">&quot;U3&quot;</span><span class="p">:</span> <span class="s2">&quot;3p_UTR_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;U5&quot;</span><span class="p">:</span> <span class="s2">&quot;5p_UTR_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TOPMED&quot;</span><span class="p">:</span> <span class="s2">&quot;AF_in_TopMed_db&quot;</span><span class="p">,</span>
    <span class="s2">&quot;WGT&quot;</span><span class="p">:</span> <span class="s2">&quot;Weight&quot;</span><span class="p">,</span>  <span class="c1"># Weight, 00 - unmapped, 1 - weight 1, 2 - weight 2, 3 - weight 3 or more</span>
    <span class="s2">&quot;MTP&quot;</span><span class="p">:</span> <span class="s2">&quot;Microattribution_TPA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LSD&quot;</span><span class="p">:</span> <span class="s2">&quot;Locus_specific_database&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NOC&quot;</span><span class="p">:</span> <span class="s2">&quot;Contig_not_in_variant_list&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DSS&quot;</span><span class="p">:</span> <span class="s2">&quot;Donor_splice_varianP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SYN&quot;</span><span class="p">:</span> <span class="s2">&quot;Synonymous_variant&quot;</span><span class="p">,</span>
    <span class="s2">&quot;KGPhase3&quot;</span><span class="p">:</span> <span class="s2">&quot;1000genome_p3&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CAF&quot;</span><span class="p">:</span> <span class="s2">&quot;Alleles_in_1000g&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VC&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_class&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MUT&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_cited&quot;</span><span class="p">,</span>
    <span class="s2">&quot;KGPhase1&quot;</span><span class="p">:</span> <span class="s2">&quot;1000genome_p1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VP&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_property&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SAO&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_Origin&quot;</span><span class="p">,</span> <span class="c1"># Origin: 0 - unspecified, 1 - Germline, 2 - Somatic, 3 - Both&quot;</span>
    <span class="s2">&quot;GENEINFO&quot;</span><span class="p">:</span> <span class="s2">&quot;Gene_info&quot;</span><span class="p">,</span>
    <span class="s2">&quot;INT&quot;</span><span class="p">:</span> <span class="s2">&quot;Is_Intronic&quot;</span><span class="p">,</span>
    <span class="s2">&quot;G5&quot;</span><span class="p">:</span> <span class="s2">&quot;MAF_Over_5pct&quot;</span><span class="p">,</span>
    <span class="s2">&quot;OM&quot;</span><span class="p">:</span> <span class="s2">&quot;Has_OMIM_OMIA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PMC&quot;</span><span class="p">:</span> <span class="s2">&quot;Pubmed&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SSR&quot;</span><span class="p">:</span> <span class="s2">&quot;Suspect_reason_code&quot;</span><span class="p">,</span> <span class="c1"># 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other&quot;</span>
    <span class="s2">&quot;RSPOS&quot;</span><span class="p">:</span> <span class="s2">&quot;Position_dbSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HD&quot;</span><span class="p">:</span> <span class="s2">&quot;High_density_genotyping_kit_dbGaP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PM&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_Precious&quot;</span><span class="p">,</span>
    <span class="s2">&quot;QD&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant_call_confidence&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MQRankSum&quot;</span><span class="p">:</span> <span class="s2">&quot;MQRankSum&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RAW_MQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Raw_Mapping_Quality_Database&quot;</span><span class="p">,</span>
    <span class="s2">&quot;FS&quot;</span><span class="p">:</span> <span class="s2">&quot;FisherStrand_bias_dbSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;POSITIVE_TRAIN_SITE&quot;</span><span class="p">:</span> <span class="s2">&quot;Used_by_GATK_VQSR_training_pos&quot;</span><span class="p">,</span>
    <span class="s2">&quot;culprit&quot;</span><span class="p">:</span> <span class="s2">&quot;Worst_annotation_VQSR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NEGATIVE_TRAIN_SITE&quot;</span><span class="p">:</span> <span class="s2">&quot;Used_by_GATK_VQSR_training_neg&quot;</span><span class="p">,</span>
    <span class="s2">&quot;variant_type&quot;</span><span class="p">:</span> <span class="s2">&quot;dbSNP_Variant_Type&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ReadPosRankSum&quot;</span><span class="p">:</span> <span class="s2">&quot;Position_bias_dbSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SOR&quot;</span><span class="p">:</span> <span class="s2">&quot;SymetricOddRatio_strand_bias_dnSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;InbreedingCoeff&quot;</span><span class="p">:</span> <span class="s2">&quot;Inbreeding_Coefficient_Hardy_Weinberg&quot;</span><span class="p">,</span>
    <span class="s2">&quot;vep&quot;</span><span class="p">:</span> <span class="s2">&quot;Vep_Annotation&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MQ&quot;</span><span class="p">:</span> <span class="s2">&quot;Mapping_Quality_dnSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;n_alt_alleles&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Alleles_dbSNP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;lcr&quot;</span><span class="p">:</span> <span class="s2">&quot;Low_Complexity_Region&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_VQSLOD&quot;</span><span class="p">:</span> <span class="s2">&quot;GATK_VQSR_log_Odds_Ratio&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf95_nfe&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_95pct_Non_Finnish_Europeans&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_asj&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Ashkenazi_Jewish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_ami_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Amish_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_fin_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Finnish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_eas&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_East_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_nfe_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Non_Finnish_Europeans_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_oth&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Other_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_ami_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Amish_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_nfe_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Non_Finnish_Europeans_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_sas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_South_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_nfe_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Non_Finnish_Europeans_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_afr&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_African_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_amr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Latino_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_amr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Latino_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_oth&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Other_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_oth_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Other_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_eas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_East_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_oth_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Other_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_nfe_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Non_Finnish_Europeans_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_nfe_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Non_Finnish_Europeans_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf95_sas&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_95pct_South_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_nfe_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Non_Finnish_Europeans_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_ami_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Amish_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_oth&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Other_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf95_afr&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_95pct_African_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_fin&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Finnish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_ami&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Amish_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_eas&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_East_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_asj&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Ashkenazi_Jewish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_amr&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Latino_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_oth_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Other_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_fin_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Finnish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_afr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_African_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_eas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_East_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_ami&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Amish_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_afr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_African_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf99_sas&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_99pct_South_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_sas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_South_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_afr&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_African_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_asj_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Ashkenazi_Jewish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_raw&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_raw&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_oth&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Other_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_oth_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Other_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_amr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Latino_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_amr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Latino_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_amr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Latino_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_sas&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_South_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_eas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_East_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_asj_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Ashkenazi_Jewish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf99_amr&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_99pct_Latino_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_oth_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Other_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_eas&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_East_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_sas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_South_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_nfe_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Non_Finnish_Europeans_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_asj_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Ashkenazi_Jewish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_asj&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Ashkenazi_Jewish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_nfe_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Non_Finnish_Europeans_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_ami_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Amish_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_sas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_South_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_afr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_African_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_nfe&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Non_Finnish_Europeans&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_ami_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Amish_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_amr&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Latino_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_nfe&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Non_Finnish_Europeans&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_fin_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Finnish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_eas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_East_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_ami_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Amish_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_eas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_East_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_eas&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_East_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf99_eas&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_99pct_East_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_fin_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Finnish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_sas&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_South_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_raw&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_raw&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_sas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_South_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_asj&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Ashkenazi_Jewish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_ami&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Amish_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf95_eas&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_95pct_East_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_nfe&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Non_Finnish_Europeans&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_afr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_African_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_fin&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Finnish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_oth_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Other_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_asj_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Ashkenazi_Jewish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_eas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_East_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_amr&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Latino_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_sas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_South_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_oth_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Other_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf95_adj&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_95pct_adj&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_eas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_East_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_fin_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Finnish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_sas&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_South_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_afr&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_African_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_ami_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Amish_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf95_amr&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_95pct_Latino_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_afr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_African_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_fin&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Finnish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_eas_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_East_Asian_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf99_nfe&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_99pct_Non_Finnish_Europeans&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_oth_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Other_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_amr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Latino_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_fin_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Finnish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_raw&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_raw&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_amr&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Latino_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_amr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Latino_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_ami_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Amish_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_asj_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Ashkenazi_Jewish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;faf99_adj&quot;</span><span class="p">:</span> <span class="s2">&quot;Filtered_Allele_Frequency_99pct_adj&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_afr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_African_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_amr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Latino_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_asj_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_Ashkenazi_Jewish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_sas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_South_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_sas_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_South_Asian_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_ami&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Amish_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_sas&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_South_Asian_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_fin_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Finnish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_asj_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_Ashkenazi_Jewish_ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_asj_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Ashkenazi_Jewish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF_afr_male&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Frequency_African_Ancestry_male&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_fin&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_Finnish_ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_fin_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_Finnish_ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nhomalt_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Homozygous_individuals_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN_afr&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_Number_African_Ancestry&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AC_afr_female&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternative_Count_African_Ancestry_female&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_NFE_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AF_NFE&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_SAS_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AF_SAS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_GERP___RS&quot;</span><span class="p">:</span> <span class="s2">&quot;GERPpp_RS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_GERP___NR&quot;</span><span class="p">:</span> <span class="s2">&quot;GERPpp_NR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_Adj_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_Adjusted_Allele_Count&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_Adj_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_Adjusted_Allele_Frequency&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_SAS_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC_SAS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_AMR_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;AMR_MAF&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_AMR_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;AMR_MAF_AC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_MetaSVM_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;MetaSNV_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_EAS_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;EAS_MAF_AC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_Interpro_domain&quot;</span><span class="p">:</span> <span class="s2">&quot;Interpro_domain&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_FATHMM_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;FATHMM_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_AFR_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AF_AFR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_AFR_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC_AFR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;1000g_general_allele_count&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;GMAF&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_EAS_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;EAS_MAF&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_AF&quot;</span><span class="p">:</span> <span class="s1">&#39;ExAC_AF&#39;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_Uniprot_acc&quot;</span><span class="p">:</span> <span class="s2">&quot;TREMBL&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_LRT_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;LRT_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_PROVEAN_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;Provean_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_FIN_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC_FIN&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_phastCons100way_vertebrate&quot;</span><span class="p">:</span> <span class="s2">&quot;phastCons100way_vertebrate_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_FIN_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AF_FIN&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_CADD_phred&quot;</span><span class="p">:</span> <span class="s2">&quot;CADD_phred_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_EUR_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;EUR_MAF_AC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_Polyphen2_HDIV_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;PolyPhen&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ESP6500_EA_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_count_Europe_American_NHLBIGO&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_AFR_AC&quot;</span><span class="p">:</span> <span class="s1">&#39;AFR_MAF_AC&#39;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_EUR_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;EUR_MAF&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_AMR_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AF_AMR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_AFR_AF&quot;</span><span class="p">:</span> <span class="s1">&#39;AFR_MAF&#39;</span><span class="p">,</span>
    <span class="s1">&#39;dbNSFP_MutationTaster_pred&#39;</span><span class="p">:</span> <span class="s2">&quot;Mutation_taster_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_MutationAssessor_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;Mutation_assessor_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ESP6500_AA_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_frequency_African_American_NHLBIGO&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_Polyphen2_HVAR_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;Polyphen_HVAR_prediction_score&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_AMR_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC_AMR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_NFE_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC_NFE&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_SIFT_pred&quot;</span><span class="p">:</span> <span class="s2">&quot;SIFT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_SAS_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;SAS_MAF_AC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_EAS_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AC_EAS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_1000Gp3_SAS_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;SAS_MAF&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ExAC_EAS_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;ExAC_AF_EAS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ESP6500_EA_AF&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_frequency_Europe_American_NHLBIGO&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dbNSFP_ESP6500_AA_AC&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele_count_African_American_NHLBIGO&quot;</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Gather parameters</span>
<span class="n">add_cols</span> <span class="o">=</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;add_cols&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">ncbi_build</span> <span class="o">=</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ncbi_build&quot;</span><span class="p">,</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">)</span>
<span class="n">center</span> <span class="o">=</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;center&quot;</span><span class="p">,</span> <span class="s2">&quot;GustaveRoussy&quot;</span><span class="p">)</span>
<span class="c1">#caller = snakemake.params.get(&quot;caller&quot;, &quot;mutect2&quot;)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parameters retrieved&quot;</span><span class="p">)</span>

<span class="c1"># Load user&#39;s data</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="n">snakemake</span><span class="o">.</span><span class="n">input</span><span class="p">[</span><span class="s2">&quot;tsv&quot;</span><span class="p">],</span>
    <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">index_col</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Variants loaded in memory&quot;</span><span class="p">)</span>

<span class="c1"># Replace header names</span>
<span class="n">new_header</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># translation_table = (</span>
<span class="c1">#     headers_mutect2</span>
<span class="c1">#     if caller.lower() == &quot;mutect2&quot;</span>
<span class="c1">#     else None</span>
<span class="c1"># )</span>
<span class="n">translation_table</span> <span class="o">=</span> <span class="n">headers</span>

<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">colname</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()):</span>
    <span class="n">new_colname</span> <span class="o">=</span> <span class="n">translation_table</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">colname</span><span class="p">,</span> <span class="n">colname</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">3</span> <span class="ow">and</span> <span class="n">colname</span> <span class="o">==</span> <span class="s2">&quot;REF&quot;</span><span class="p">:</span>
        <span class="n">new_header</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reference_Allele&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">new_header</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_colname</span> <span class="k">if</span> <span class="n">new_colname</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">colname</span><span class="p">)</span>
<span class="n">variants</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">new_header</span>
<span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;New header defined&quot;</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>

<span class="c1"># Add new columns on demand</span>
<span class="k">if</span> <span class="n">add_cols</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
    <span class="k">if</span> <span class="s2">&quot;Center&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding Sequencing Center information&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Center&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;NCBI_Build&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding NCBI build information&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;NCBI_Build&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;End_Position&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding End position&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">start</span> <span class="o">-</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">ref</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">],</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Reference_Allele&quot;</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;Variant_Type&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding variant type&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Variant_Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">get_variant_type</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span>
            <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Reference_Allele&quot;</span><span class="p">],</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Tumor_Seq_Allele1&quot;</span><span class="p">])</span>
        <span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;SYMBOL&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding gene symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;HGNC_ID&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding hugo symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;HGNC_ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_AFR&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_African_Ancestry&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_AFR symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_AFR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_African_Ancestry&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_AMR&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_Latino_Ancestry&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_AMR symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_AMR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_Latino_Ancestry&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_EAS&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_East_Asian_ancestry&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_EAS symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_EAS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_East_Asian_ancestry&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_FIN&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_Finnish_ancestry&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_FIN symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_FIN&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_Finnish_ancestry&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_NFE&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_Non_Finnish_Europeans&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_NFE symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_NFE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_Non_Finnish_Europeans&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_OTH&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_Other_Ancestry&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_OTH symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_OTH&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_Other_Ancestry&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;ExAC_AF_SAS&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;Allele_Frequency_South_Asian_ancestry&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding ExAC_AF_SAS symbols&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;ExAC_AF_SAS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Allele_Frequency_South_Asian_ancestry&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;vcf_region&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Adding VCF regions&quot;</span><span class="p">)</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;vcf_region&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="p">[</span><span class="nb">chr</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">]))</span>
            <span class="k">for</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">],</span>
                <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">],</span>
                <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Variant_ID&quot;</span><span class="p">],</span>
                <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Reference_Allele&quot;</span><span class="p">],</span>
                <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Tumor_Seq_Allele1&quot;</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;Variant_Origin&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Translating Variant_Origin&quot;</span><span class="p">)</span>
        <span class="n">translation</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;0&quot;</span><span class="p">:</span> <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
            <span class="s2">&quot;0.0&quot;</span><span class="p">:</span> <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nan&quot;</span><span class="p">:</span> <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
            <span class="s2">&quot;1&quot;</span><span class="p">:</span> <span class="s2">&quot;germline&quot;</span><span class="p">,</span>
            <span class="s2">&quot;1.0&quot;</span><span class="p">:</span> <span class="s2">&quot;germline&quot;</span><span class="p">,</span>
            <span class="s2">&quot;2&quot;</span><span class="p">:</span> <span class="s2">&quot;somatic&quot;</span><span class="p">,</span>
            <span class="s2">&quot;2.0&quot;</span><span class="p">:</span> <span class="s2">&quot;somatic&quot;</span><span class="p">,</span>
            <span class="s2">&quot;3&quot;</span><span class="p">:</span> <span class="s2">&quot;both_germline_somatic&quot;</span><span class="p">,</span>
            <span class="s2">&quot;3.0&quot;</span><span class="p">:</span> <span class="s2">&quot;both_germline_somatic&quot;</span>
        <span class="p">}</span>

        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Variant_Origin_Readable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">translation</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Variant_Origin&quot;</span><span class="p">]</span>
        <span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;Suspect_reason_code&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">translation</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;0&quot;</span><span class="p">:</span> <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
            <span class="s2">&quot;1&quot;</span><span class="p">:</span> <span class="s2">&quot;Paralog&quot;</span><span class="p">,</span>
            <span class="s2">&quot;2&quot;</span><span class="p">:</span> <span class="s2">&quot;byEST&quot;</span><span class="p">,</span>
            <span class="s2">&quot;4&quot;</span><span class="p">:</span> <span class="s2">&quot;oldAlign&quot;</span><span class="p">,</span>
            <span class="s2">&quot;8&quot;</span><span class="p">:</span> <span class="s2">&quot;Para_EST&quot;</span><span class="p">,</span>
            <span class="s2">&quot;16&quot;</span><span class="p">:</span> <span class="s2">&quot;1kg_failed&quot;</span><span class="p">,</span>
            <span class="s2">&quot;1024&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
            <span class="s2">&quot;0.0&quot;</span><span class="p">:</span> <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
            <span class="s2">&quot;1.0&quot;</span><span class="p">:</span> <span class="s2">&quot;Paralog&quot;</span><span class="p">,</span>
            <span class="s2">&quot;2.0&quot;</span><span class="p">:</span> <span class="s2">&quot;byEST&quot;</span><span class="p">,</span>
            <span class="s2">&quot;4.0&quot;</span><span class="p">:</span> <span class="s2">&quot;oldAlign&quot;</span><span class="p">,</span>
            <span class="s2">&quot;8.0&quot;</span><span class="p">:</span> <span class="s2">&quot;Para_EST&quot;</span><span class="p">,</span>
            <span class="s2">&quot;16.0&quot;</span><span class="p">:</span> <span class="s2">&quot;1kg_failed&quot;</span><span class="p">,</span>
            <span class="s2">&quot;1024.0&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nan&quot;</span><span class="p">:</span> <span class="s2">&quot;unspecified&quot;</span>
        <span class="p">}</span>

        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Suspect_reason_code_Readable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">translation</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Suspect_reason_code&quot;</span><span class="p">]</span>
        <span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;Tumor_Sample_Barcode&quot;</span> <span class="ow">in</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span>
        <span class="p">]</span>

    <span class="k">if</span> <span class="s2">&quot;Matched_Norm_Sample_Barcode&quot;</span> <span class="ow">in</span> <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Matched_Norm_Sample_Barcode&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">snakemake</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s2">&quot;Matched_Norm_Sample_Barcode&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span>
        <span class="p">]</span>

<span class="n">variants</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">snakemake</span><span class="o">.</span><span class="n">output</span><span class="p">[</span><span class="s2">&quot;tsv&quot;</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="split_design.html" class="btn btn-neutral float-right" title="SPLIT_DESIGN" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="multiqc_rnaseq_report.html" class="btn btn-neutral float-left" title="MULTIQC_RNASEQ_REPORT" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2016, The Snakemake team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>