

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>SNPEFF_SNPSIFT (UNDER DEVELOPMENT) &mdash; Snakemake Wrappers heads/Unofficial documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="VARIANT_CALLING_AMPLISEQ" href="variant_calling_ampliseq.html" />
    <link rel="prev" title="SALMON_QUANT" href="salmon_quant.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Snakemake Wrappers
          

          
          </a>

          
            
            
              <div class="version">
                heads/Unofficial
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Wrappers, Meta-Wrappers and Pipelines</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../meta-wrappers.html">Meta-Wrappers</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../pipelines.html">BiGR Pipelines</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="WES_germline.html">WES_GERMLINE</a></li>
<li class="toctree-l2"><a class="reference internal" href="bcl2fastq.html">BCL2FASTQ (UNDER DEVELOPMENT)</a></li>
<li class="toctree-l2"><a class="reference internal" href="celltype_deconvolution_rnaseq.html">CELLTYPE_DECONVOLUTION_RNASEQ (UNDER DEVELOPMENT)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cytoscan_eacon.html">EACON_CYTOSCAN</a></li>
<li class="toctree-l2"><a class="reference internal" href="design_builder.html">DESIGN_BUILDER</a></li>
<li class="toctree-l2"><a class="reference internal" href="dge_deseq2.html">SALMON_QUANT</a></li>
<li class="toctree-l2"><a class="reference internal" href="fastqc_multiqc.html">FASTQC_MULTIQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="oncoscan_eacon.html">EACON_ONCOSCAN</a></li>
<li class="toctree-l2"><a class="reference internal" href="salmon_quant.html">SALMON_QUANT</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">SNPEFF_SNPSIFT (UNDER DEVELOPMENT)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#input-output">Input/Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#used-meta-wrappers">Used meta-wrappers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#used-wrappers">Used wrappers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#snakefile">Snakefile</a></li>
<li class="toctree-l3"><a class="reference internal" href="#authors">Authors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="variant_calling_ampliseq.html">VARIANT_CALLING_AMPLISEQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_calling_germline.html">VARIANT_CALLING_GERMLINE (UNDER DEVELOPMENT)</a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_calling_germline_wes.html">VARIANT_CALLING_GERMLINE_WES (UNDER DEVELOPMENT)</a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_calling_somatic.html">VARIANT_CALLING_SOMATIC (UNDER DEVELOPMENT)</a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_callling_germline.html">VARIANT_CALLING_GERMLINE (UNDER DEVELOPMENT)</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Snakemake Wrappers</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../pipelines.html">BiGR Pipelines</a> &raquo;</li>
        
      <li>SNPEFF_SNPSIFT (UNDER DEVELOPMENT)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/bigr-pipelines/snpeff_snpsift.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="snpeff-snpsift-under-development">
<span id="id1"></span><h1>SNPEFF_SNPSIFT (UNDER DEVELOPMENT)<a class="headerlink" href="#snpeff-snpsift-under-development" title="Permalink to this headline">¶</a></h1>
<p>Annotate VCF files with SnpEff and SNpSift</p>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>In order to run the pipeline, use the following commands</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Go to your directory containing VCF files</span>

<span class="nb">cd</span> /path/to/VCF/dir

<span class="c1"># Copy/paste the following line for **HG38**</span>

bash /mnt/beegfs/pipelines/snakemake-wrappers/bigr_pipelines/snpeff_snpsift/run.sh

<span class="c1"># Copy/paste the following line for **HG19**</span>

bash /mnt/beegfs/pipelines/snakemake-wrappers/bigr_pipelines/snpeff_snpsift/run.sh hg19
</pre></div>
</div>
</div>
<div class="section" id="input-output">
<h2>Input/Output<a class="headerlink" href="#input-output" title="Permalink to this headline">¶</a></h2>
<p><strong>Input:</strong></p>
<ul class="simple">
<li><p>VCF files</p></li>
</ul>
<p><strong>Output:</strong></p>
<ul class="simple">
<li><p>Annotated VCF file</p></li>
</ul>
</div>
<div class="section" id="used-meta-wrappers">
<h2>Used meta-wrappers<a class="headerlink" href="#used-meta-wrappers" title="Permalink to this headline">¶</a></h2>
<p>The following individual meta-wrappers are used in this pipeline:</p>
<ul class="simple">
<li><p><span class="xref std std-ref">meta/bio/snpeff_annotate</span></p></li>
<li><p><span class="xref std std-ref">meta/bio/snpsift</span></p></li>
</ul>
<p>Please refer to each meta-wrapper in above list for additional configuration parameters and information about the executed code.</p>
</div>
<div class="section" id="used-wrappers">
<h2>Used wrappers<a class="headerlink" href="#used-wrappers" title="Permalink to this headline">¶</a></h2>
<p>The following individual wrappers are used in this pipeline:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../wrappers/tabix.html#bio-tabix"><span class="std std-ref">TABIX</span></a></p></li>
<li><p><a class="reference internal" href="../wrappers/compress/pbgzip.html#bio-compress-pbgzip"><span class="std std-ref">COMPRESS_PBGZIP</span></a></p></li>
<li><p><a class="reference internal" href="../wrappers/multiqc.html#bio-multiqc"><span class="std std-ref">MULTIQC</span></a></p></li>
</ul>
<p>Please refer to each wrapper in above list for additional configuration parameters and information about the executed code.</p>
</div>
<div class="section" id="snakefile">
<h2>Snakefile<a class="headerlink" href="#snakefile" title="Permalink to this headline">¶</a></h2>
<p>The pipeline contains the following steps:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="n">worflow_source_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">workflow</span><span class="o">.</span><span class="n">get_sources</span><span class="p">())))</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span><span class="o">.</span><span class="n">parent</span>
<span class="n">common</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">worflow_source_dir</span> <span class="o">/</span> <span class="s2">&quot;../common/python&quot;</span><span class="p">)</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">common</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">file_manager</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">files_linker</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">write_yaml</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">messages</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">snakemake.utils</span> <span class="kn">import</span> <span class="n">min_version</span>
<span class="n">min_version</span><span class="p">(</span><span class="s2">&quot;6.0&quot;</span><span class="p">)</span>

<span class="n">default_config</span> <span class="o">=</span> <span class="n">read_yaml</span><span class="p">(</span><span class="n">worflow_source_dir</span> <span class="o">/</span> <span class="s2">&quot;config.hg38.yaml&quot;</span><span class="p">)</span>
<span class="n">configfile</span><span class="p">:</span> <span class="n">get_config</span><span class="p">(</span><span class="n">default_config</span><span class="p">)</span>
<span class="n">design</span> <span class="o">=</span> <span class="n">build_design</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="n">search_fastq_pairs</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>


<span class="n">samples_list</span> <span class="o">=</span> <span class="n">design</span><span class="p">[</span><span class="s2">&quot;Sample_id&quot;</span><span class="p">]</span>


<span class="n">wildcard_constraints</span><span class="p">:</span>
    <span class="n">sample</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">samples_list</span><span class="p">)</span>


<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">calls</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span>
            <span class="s2">&quot;snpsift/dbnsfp/</span><span class="si">{sample}</span><span class="s2">.vcf.gz</span><span class="si">{index}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">sample</span><span class="o">=</span><span class="n">samples_list</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;.tbi&quot;</span><span class="p">]</span>
        <span class="p">),</span>
        <span class="n">qc</span><span class="o">=</span><span class="s2">&quot;multiqc/SnpEff_annotation.html&quot;</span><span class="p">,</span>
        <span class="n">tsv</span><span class="o">=</span><span class="n">expand</span><span class="p">(</span>
            <span class="s2">&quot;snpsift/extractFields/</span><span class="si">{sample}</span><span class="s2">.tsv&quot;</span><span class="p">,</span>
            <span class="n">sample</span><span class="o">=</span><span class="n">samples_list</span>
        <span class="p">)</span>
    <span class="n">message</span><span class="p">:</span>
        <span class="s2">&quot;Finishing the annotation pipeline&quot;</span>

<span class="c1">#####################</span>
<span class="c1">### Export to TSV ###</span>
<span class="c1">#####################</span>

<span class="n">rule</span> <span class="n">extractfields</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">call</span><span class="o">=</span><span class="s2">&quot;snpsift/cosmic/</span><span class="si">{sample}</span><span class="s2">.vcf.gz&quot;</span><span class="p">,</span>
        <span class="n">call_index</span><span class="o">=</span><span class="s2">&quot;snpsift/cosmic/</span><span class="si">{sample}</span><span class="s2">.vcf.gz.tbi&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">tsv</span><span class="o">=</span><span class="n">protected</span><span class="p">(</span><span class="s2">&quot;snpsift/extractFields/</span><span class="si">{sample}</span><span class="s2">.tsv&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">:</span>
        <span class="s2">&quot;Making </span><span class="si">{wildcards.sample}</span><span class="s2"> annotated VCF readable&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">2</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">attempt</span> <span class="o">*</span> <span class="mi">4096</span><span class="p">,</span> <span class="mi">15360</span><span class="p">),</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">20</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/snpsift/extractFields/</span><span class="si">{sample}</span><span class="s2">.log&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">fields</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;CHROM&quot;</span><span class="p">,</span>
            <span class="s2">&quot;POS&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;REF&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ALT&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ANN[*].GENE&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ANN[*].GENEID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ANN[*].HGVS_C&quot;</span><span class="p">,</span> <span class="c1"># (alias HGVS_DNA, CODON): Variant in HGVS (DNA) notation</span>
            <span class="s2">&quot;ANN[*].HGVS_P&quot;</span><span class="p">,</span> <span class="c1"># (alias HGVS, HGVS_PROT, AA): Variant in HGVS (protein) notation</span>
            <span class="s2">&quot;STRAND&quot;</span><span class="p">,</span> <span class="c1"># &quot;Gene strand&quot;</span>
            <span class="s2">&quot;ANN[*].FEATURE&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ANN[*].FEATUREID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ANN[*].BIOTYPE&quot;</span><span class="p">,</span> <span class="c1"># Biotype, as described by the annotations (e.g. &#39;protein_coding&#39;)</span>
            <span class="s2">&quot;ANN[*].IMPACT&quot;</span><span class="p">,</span> <span class="c1"># { HIGH, MODERATE, LOW, MODIFIER }</span>
            <span class="s2">&quot;ANN[*].EFFECT&quot;</span><span class="p">,</span> <span class="c1"># (alias ANNOTATION): Effect in Sequence ontology terms (e.g. &#39;missense_variant&#39;, &#39;synonymous_variant&#39;, &#39;stop_gained&#39;, etc.)</span>
            <span class="s2">&quot;ANN[*].CDNA_POS&quot;</span><span class="p">,</span> <span class="c1"># (alias POS_CDNA)</span>
            <span class="s2">&quot;ANN[*].CDNA_LEN&quot;</span><span class="p">,</span> <span class="c1"># (alias LEN_CDNA)</span>
            <span class="s2">&quot;ANN[*].CDS_POS&quot;</span><span class="p">,</span> <span class="c1"># (alias POS_CDS)</span>
            <span class="s2">&quot;ANN[*].CDS_LEN&quot;</span><span class="p">,</span> <span class="c1"># (alias LEN_CDS)</span>
            <span class="s2">&quot;ANN[*].AA_POS&quot;</span><span class="p">,</span> <span class="c1"># (alias POS_AA)</span>
            <span class="s2">&quot;ANN[*].AA_LEN&quot;</span><span class="p">,</span> <span class="c1"># (alias LEN_AA)</span>
            <span class="s2">&quot;FILTER&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ADP&quot;</span><span class="p">,</span> <span class="c1"># &quot;Average per-sample depth of bases with Phred score &gt;= 15&quot;</span>
            <span class="s2">&quot;LOF[*].GENE&quot;</span><span class="p">,</span>  <span class="c1"># predicted Loss of function</span>
            <span class="s2">&quot;LOF[*].GENEID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;LOF[*].NUMTR&quot;</span><span class="p">,</span> <span class="c1"># Number of transcripts affected</span>
            <span class="s2">&quot;LOF[*].PERC&quot;</span><span class="p">,</span> <span class="c1"># percentage of transcripts</span>
            <span class="s2">&quot;NMD[*].GENE&quot;</span><span class="p">,</span> <span class="c1"># &quot;Predicted nonsense mediated decay effects for this variant.&quot;</span>
            <span class="s2">&quot;NMD[*].GENEID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;NMD[*].NUMTR&quot;</span><span class="p">,</span>
            <span class="s2">&quot;NMD[*].PERC&quot;</span><span class="p">,</span>
            <span class="s2">&quot;VARTYPE&quot;</span><span class="p">,</span> <span class="c1"># &quot;Comma separated list of variant types. One per allele&quot;</span>
            <span class="s2">&quot;HOM&quot;</span><span class="p">,</span> <span class="c1"># &quot;Variant is homozygous&quot;</span>
            <span class="s2">&quot;HET&quot;</span><span class="p">,</span> <span class="c1"># &quot;Variant is heterozygous&quot;</span>
            <span class="s2">&quot;AC&quot;</span><span class="p">,</span> <span class="c1"># &quot;Allele Count&quot;</span>
            <span class="s2">&quot;AF&quot;</span><span class="p">,</span> <span class="c1"># &quot;Allele Frequency&quot;</span>
            <span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="c1">#&quot;Total number of alleles in data sources&quot;</span>
            <span class="s2">&quot;DS&quot;</span><span class="p">,</span> <span class="c1"># &quot;Data Sources containing allele&quot;</span>
            <span class="s2">&quot;CDA&quot;</span><span class="p">,</span> <span class="c1"># &quot;Variation is interrogated in a clinical diagnostic assay&quot;</span>
            <span class="s2">&quot;OTH&quot;</span><span class="p">,</span> <span class="c1"># &quot;Has other variant with exactly the same set of mapped positions on NCBI refernce assembly.&quot;</span>
            <span class="s2">&quot;COMMON&quot;</span><span class="p">,</span> <span class="c1"># A common SNP is one that has at least one 1000Genomes population with a minor allel e of frequency &gt;= 1% and for which 2 or more founders contribute to that minor allele frequency.</span>
            <span class="s2">&quot;VLD&quot;</span><span class="p">,</span> <span class="c1"># &quot;Is Validated.  This bit is set if the variant has 2+ minor allele count based on frequency or genotype data.&quot;</span>
            <span class="s2">&quot;MUT&quot;</span><span class="p">,</span> <span class="c1"># &quot;Is mutation (journal citation, explicit fact): a low frequency variation that is cited in journal and other reputable sources&quot;</span>
            <span class="s2">&quot;INT&quot;</span><span class="p">,</span> <span class="c1"># &quot;In Intron&quot;</span>
            <span class="s2">&quot;PMC&quot;</span><span class="p">,</span> <span class="c1"># &quot;Links exist to PubMed Central article&quot;</span>
            <span class="s2">&quot;HD&quot;</span><span class="p">,</span> <span class="c1"># &quot;Marker is on high density genotyping kit (50K density or greater).  The variant may have phenotype associations present in dbGaP.&quot;</span>
            <span class="s2">&quot;PM&quot;</span><span class="p">,</span> <span class="c1"># &quot;Variant is Precious(Clinical,Pubmed Cited)&quot;</span>
        <span class="p">],</span>
        <span class="n">extra</span><span class="o">=</span><span class="s2">&quot;-s &#39;</span><span class="se">\t</span><span class="s2">&#39; -e &#39;.&#39;&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/snpsift/extractfields&quot;</span>


<span class="c1">###############</span>
<span class="c1">### MultiQC ###</span>
<span class="c1">###############</span>

<span class="n">rule</span> <span class="n">multiqc</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span>
            <span class="s2">&quot;snpeff/report/</span><span class="si">{sample}</span><span class="s2">.html&quot;</span><span class="p">,</span>
            <span class="n">sample</span><span class="o">=</span><span class="n">samples_list</span>
        <span class="p">),</span>
        <span class="n">expand</span><span class="p">(</span>
            <span class="s2">&quot;snpeff/csvstats/</span><span class="si">{sample}</span><span class="s2">.csv&quot;</span><span class="p">,</span>
            <span class="n">sample</span><span class="o">=</span><span class="n">samples_list</span>
        <span class="p">)</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">report</span><span class="p">(</span>
            <span class="s2">&quot;multiqc/SnpEff_annotation.html&quot;</span><span class="p">,</span>
            <span class="n">caption</span><span class="o">=</span><span class="s2">&quot;../common/reports/multiqc.rst&quot;</span><span class="p">,</span>
            <span class="n">category</span><span class="o">=</span><span class="s2">&quot;Quality Controls&quot;</span>
        <span class="p">)</span>
    <span class="n">message</span><span class="p">:</span>
        <span class="s2">&quot;Aggregating quality reports from SnpEff&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">attempt</span> <span class="o">*</span> <span class="mi">1536</span><span class="p">,</span> <span class="mi">10240</span><span class="p">),</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">35</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/multiqc.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/multiqc&quot;</span>


<span class="c1">#################################</span>
<span class="c1">### FINAL VCF FILE INDEXATION ###</span>
<span class="c1">#################################</span>

<span class="n">rule</span> <span class="n">tabix_index</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;</span><span class="si">{tools}</span><span class="s2">/</span><span class="si">{subcommand}</span><span class="s2">/</span><span class="si">{sample}</span><span class="s2">.vcf.gz&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;</span><span class="si">{tools}</span><span class="s2">/</span><span class="si">{subcommand}</span><span class="s2">/</span><span class="si">{sample}</span><span class="s2">.vcf.gz.tbi&quot;</span>
    <span class="n">message</span><span class="p">:</span>
        <span class="s2">&quot;Indexing </span><span class="si">{wildcards.sample}</span><span class="s2"> final annotated VCF with tabix&quot;</span>
    <span class="n">group</span><span class="p">:</span>
        <span class="s2">&quot;Final_compression_</span><span class="si">{sample}</span><span class="s2">&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">1020</span><span class="p">,</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">45</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/</span><span class="si">{tools}</span><span class="s2">/</span><span class="si">{subcommand}</span><span class="s2">/tabix/</span><span class="si">{sample}</span><span class="s2">.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/tabix&quot;</span>


<span class="n">rule</span> <span class="n">compress_pbgzip</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s2">&quot;</span><span class="si">{tools}</span><span class="s2">/</span><span class="si">{subcommand}</span><span class="s2">/</span><span class="si">{sample}</span><span class="s2">.vcf&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;</span><span class="si">{tools}</span><span class="s2">/</span><span class="si">{subcommand}</span><span class="s2">/</span><span class="si">{sample}</span><span class="s2">.vcf.gz&quot;</span>
    <span class="n">message</span><span class="p">:</span>
        <span class="s2">&quot;Compressing </span><span class="si">{wildcards.sample}</span><span class="s2"> final annotated VCF with pbgzip&quot;</span>
    <span class="n">group</span><span class="p">:</span>
        <span class="s2">&quot;Final_compression_</span><span class="si">{sample}</span><span class="s2">&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="n">resources</span><span class="p">:</span>
        <span class="n">mem_mb</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">1020</span><span class="p">,</span>
        <span class="n">time_min</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">,</span> <span class="n">attempt</span><span class="p">:</span> <span class="n">attempt</span> <span class="o">*</span> <span class="mi">45</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;logs/</span><span class="si">{tools}</span><span class="s2">/</span><span class="si">{subcommand}</span><span class="s2">/pbgzip/</span><span class="si">{sample}</span><span class="s2">.log&quot;</span>
    <span class="n">wrapper</span><span class="p">:</span>
        <span class="s2">&quot;bio/compress/pbgzip&quot;</span>


<span class="c1">######################</span>
<span class="c1">### VCF annotation ###</span>
<span class="c1">######################</span>


<span class="n">module</span> <span class="n">snpeff_meta</span><span class="p">:</span>
    <span class="n">snakefile</span><span class="p">:</span> <span class="s2">&quot;../../meta/bio/snpeff_annotate/test/Snakefile&quot;</span>
    <span class="n">config</span><span class="p">:</span> <span class="n">config</span>

<span class="n">use</span> <span class="n">rule</span> <span class="n">snpeff</span> <span class="kn">from</span> <span class="nn">snpeff_meta</span> <span class="k">with</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">calls</span><span class="o">=</span><span class="s2">&quot;calls/</span><span class="si">{sample}</span><span class="s2">.vcf.gz&quot;</span><span class="p">,</span>
        <span class="n">calls_index</span><span class="o">=</span><span class="s2">&quot;calls/</span><span class="si">{sample}</span><span class="s2">.vcf.gz.tbi&quot;</span><span class="p">,</span>
        <span class="n">db</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">][</span><span class="s2">&quot;snpeff&quot;</span><span class="p">]</span>


<span class="n">module</span> <span class="n">snpsift</span><span class="p">:</span>
    <span class="n">snakefile</span><span class="p">:</span> <span class="s2">&quot;../../meta/bio/snpsift/test/Snakefile&quot;</span>
    <span class="n">config</span><span class="p">:</span> <span class="n">config</span>

<span class="n">use</span> <span class="n">rule</span> <span class="o">*</span> <span class="kn">from</span> <span class="nn">snpsift</span> <span class="k">as</span> <span class="n">snpsift_</span><span class="o">*</span>
</pre></div>
</div>
</div>
<div class="section" id="authors">
<h2>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Thibault Dayris</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="variant_calling_ampliseq.html" class="btn btn-neutral float-right" title="VARIANT_CALLING_AMPLISEQ" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="salmon_quant.html" class="btn btn-neutral float-left" title="SALMON_QUANT" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2016, The Snakemake team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>